<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • orthologr</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/drostlab/orthologr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/drostlab/orthologr/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="0.4.1" id="orthologr-version-041">
<code>orthologr</code> version 0.4.1<a class="anchor" aria-label="anchor" href="#orthologr-version-041"></a></h2>
<div class="section level3">
<h3 id="new-features-0-4-1">New Features<a class="anchor" aria-label="anchor" href="#new-features-0-4-1"></a></h3>
<ul><li>the <code><a href="../reference/divergence_stratigraphy.html">divergence_stratigraphy()</a></code> and <code><a href="../reference/divergence_map.html">divergence_map()</a></code> functions now include the parameter <code>n_quantile</code>, which enables users to choose the number of quantiles to generate for the <code>divergence map</code>. This could allow users to get a higher-resolution <code>divergence map</code> if <code>n_quantile</code> is greater than 10. Alternatively, this can resolve the issue of empty divergence strata when deciling the dNdS values for closely related organisms with dNdS = 0 for over 10% of the genes.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.4.0" id="orthologr-version-040">
<code>orthologr</code> version 0.4.0<a class="anchor" aria-label="anchor" href="#orthologr-version-040"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-4-0">New Functions<a class="anchor" aria-label="anchor" href="#new-functions-0-4-0"></a></h3>
<ul><li>new function <code><a href="../reference/check_annotation.html">check_annotation()</a></code> helps to detect corrupt GFF or GTF annotation files and removes such outlier lines</li>
</ul></div>
<div class="section level3">
<h3 id="new-features-0-4-0">New Features<a class="anchor" aria-label="anchor" href="#new-features-0-4-0"></a></h3>
<ul><li><p>the <code><a href="../reference/generate_ortholog_tables.html">generate_ortholog_tables()</a></code> and <code><a href="../reference/retrieve_longest_isoforms.html">retrieve_longest_isoforms()</a></code> now include the new <code><a href="../reference/check_annotation.html">check_annotation()</a></code> function to capture corrupt GFF or GTF files and fix them</p></li>
<li><p>adding a new argument <code>of_path</code> to <code><a href="../reference/orthofinder2.html">orthofinder2()</a></code> to allow users to specify their own path to their locally installed <code>orthofinder</code> executable</p></li>
<li><p>adding new argument <code>task</code> to <code><a href="../reference/map_generator_lnc.html">map_generator_lnc()</a></code> and <code><a href="../reference/orthologs_lnc.html">orthologs_lnc()</a></code> to allow users to use the full <code>blastn</code> range provided by <a href="https://drostlab.github.io/metablastr/reference/blast_nucleotide_to_nucleotide.html" class="external-link">blast_nucleotide_to_nucleotide()</a></p></li>
<li><p>adding new argument <code>path</code> to <code><a href="../reference/map_generator_lnc.html">map_generator_lnc()</a></code> to allow users to specify their local installation path of BLAST</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.3.0" id="orthologr-version-030">
<code>orthologr</code> version 0.3.0<a class="anchor" aria-label="anchor" href="#orthologr-version-030"></a></h2>
<div class="section level3">
<h3 id="new-functions-0-3-0">New Functions<a class="anchor" aria-label="anchor" href="#new-functions-0-3-0"></a></h3>
<ul><li><p>new function <code><a href="../reference/plot_pairwise_orthologs.html">plot_pairwise_orthologs()</a></code> allows users to plot pairwise orthologs tables for multiple pairwise comparisons</p></li>
<li><p>new function <code><a href="../reference/retrieve_core_orthologs.html">retrieve_core_orthologs()</a></code> allows users to retrieve a core set of orthologous gene loci from several pairwise ortholog tables</p></li>
<li><p>new functions <code><a href="../reference/generate_ortholog_tables.html">generate_ortholog_tables()</a></code> and <code><a href="../reference/generate_ortholog_tables_all.html">generate_ortholog_tables_all()</a></code> allow users to generate ortholog tables by gene locus and splice varaint for a set of species</p></li>
<li><p>new function <code><a href="../reference/retrieve_longest_isoforms_all.html">retrieve_longest_isoforms_all()</a></code> allows users to specify folders and retrieve the longest splice variants for all proteomes stored in a folder</p></li>
<li><p>new functions <code><a href="../reference/translate_cds_to_protein.html">translate_cds_to_protein()</a></code> and <code><a href="../reference/translate_cds_to_protein_all.html">translate_cds_to_protein_all()</a></code> which translate coding sequences into amino acid sequences for single or multiple files</p></li>
</ul></div>
<div class="section level3">
<h3 id="new-features-0-3-0">New Features<a class="anchor" aria-label="anchor" href="#new-features-0-3-0"></a></h3>
<ul><li><p>in <code><a href="../reference/orthologs.html">orthologs()</a></code> the default value of <code>delete_corrupt_cds</code> changed from <code>delete_corrupt_cds = TRUE</code> to <code>delete_corrupt_cds = FALSE</code> to be consistent with <code><a href="../reference/dNdS.html">dNdS()</a></code> and <code><a href="../reference/divergence_stratigraphy.html">divergence_stratigraphy()</a></code></p></li>
<li><p>the <code><a href="../reference/divergence_stratigraphy.html">divergence_stratigraphy()</a></code> function received a new argument <code>dnds_est.method</code> which now allows users to select different dNdS estimation methods when running <code><a href="../reference/divergence_stratigraphy.html">divergence_stratigraphy()</a></code> (suggested by Momir Futo)</p></li>
<li><p>the <code><a href="../reference/divergence_stratigraphy.html">divergence_stratigraphy()</a></code> function now allows to change the <code>eval</code> argument which wasn’t passed down to the <code><a href="../reference/dNdS.html">dNdS()</a></code> call within the function (Many thanks to Momir Futo)</p></li>
<li><p>the function <code>map.generator()</code> was renamed to <code><a href="../reference/map_generator_dnds.html">map_generator_dnds()</a></code> to be more consistent with the notation of other functions</p></li>
<li><p>the function <code>map.generator.lnc()</code> was renamed to <code><a href="../reference/map_generator_lnc.html">map_generator_lnc()</a></code> to be more consistent with the notation of other functions</p></li>
<li><p>the function <code>DivergenceMap()</code> was renamed to <code><a href="../reference/divergence_map.html">divergence_map()</a></code> to be more consistent with the notation of other functions</p></li>
<li><p>the function <code>DivergenceMap()</code> was renamed to <code><a href="../reference/divergence_map.html">divergence_map()</a></code> to be more consistent with the notation of other functions</p></li>
<li><p>the function <code>orthologs.lnc</code> was renamed to <code>orthologs_lnc</code> to be more consistent with the notation of other functions</p></li>
<li><p>the function <code>OF2CoreOrthologs()</code> was renamed to <code><a href="../reference/orthofinder2_retrieve_core_orthologs.html">orthofinder2_retrieve_core_orthologs()</a></code></p></li>
</ul></div>
<div class="section level3">
<h3 id="removed-functions-0-3-0">Removed Functions<a class="anchor" aria-label="anchor" href="#removed-functions-0-3-0"></a></h3>
<ul><li><p>the function <code>advanced_blast()</code> is not supported anymore and thus is not available to users anymore (please consult the <a href="https://github.com/HajkD/metablastr" class="external-link">metablastr package</a> in case you need this functionality)</p></li>
<li><p>the function <code>advanced_makedb()</code> is not supported anymore and thus is not available to users anymore (please consult the <a href="https://github.com/HajkD/metablastr" class="external-link">metablastr package</a> in case you need this functionality)</p></li>
<li><p>the function <code>blast.nr()</code> is not supported anymore and thus is not available to users anymore (please consult the <a href="https://github.com/HajkD/metablastr" class="external-link">metablastr package</a> in case you need this functionality)</p></li>
<li><p>the function <code>delta.blast()</code> is not supported anymore and thus is not available to users anymore (please consult the <a href="https://github.com/HajkD/metablastr" class="external-link">metablastr package</a> in case you need this functionality)</p></li>
<li><p>the function <code>ProteinOrtho()</code> is not supported anymore and thus is not available to users anymore</p></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.0" id="orthologr-version-020">
<code>orthologr</code> version 0.2.0<a class="anchor" aria-label="anchor" href="#orthologr-version-020"></a></h2>
<div class="section level3">
<h3 id="new-features-0-2-0">New Features<a class="anchor" aria-label="anchor" href="#new-features-0-2-0"></a></h3>
<ul><li>new function <code><a href="../reference/retrieve_longest_isoforms.html">retrieve_longest_isoforms()</a></code> which enables retrieval of the longest isoforms from a proteome file and save results as fasta file for downstream analyses</li>
<li>new function <code>OF2CoreOrthologs()</code> to retrieve core orthologs across multiple species from Orthofinder2 output</li>
<li>new function <code><a href="../reference/extract_features.html">extract_features()</a></code>: Helper function to extract gene loci and splice variant IDs from GFF files</li>
<li>new function <code><a href="../reference/filter_best_hits.html">filter_best_hits()</a></code>: Helper function to select best BLAST hit based on minimum evalue</li>
<li>new function <code><a href="../reference/generate_ortholog_tables.html">generate_ortholog_tables()</a></code>: Generate ortholog tables by gene locus and splice varaint</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.0" id="orthologr-version-010">
<code>orthologr</code> version 0.1.0<a class="anchor" aria-label="anchor" href="#orthologr-version-010"></a></h2>
<div class="section level3">
<h3 id="new-features-0-1-0">New Features<a class="anchor" aria-label="anchor" href="#new-features-0-1-0"></a></h3>
<ul><li>
<code><a href="../reference/read.cds.html">read.cds()</a></code> now trimms corrupted CDS (= CDS not divisible by 3) when <code>delete_corrupt_cds = FALSE</code> is specified</li>
<li>the new default value of argument <code>delete_corrupt_cds</code> in <code><a href="../reference/dNdS.html">dNdS()</a></code> is now <code>FALSE</code>. Thus, given the new trimming feature in <code><a href="../reference/read.cds.html">read.cds()</a></code>, corrupted CDS equences will be trimmed before being translated.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.5" id="orthologr-version-005">
<code>orthologr</code> version 0.0.5<a class="anchor" aria-label="anchor" href="#orthologr-version-005"></a></h2>
<div class="section level3">
<h3 id="bug-fixes-0-0-5">Bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-0-5"></a></h3>
<ul><li>The default setting of the <code>BLAST</code> argument <code>max_target_seqs 1</code> was removed from <code><a href="../reference/blast_best.html">blast_best()</a></code> and <code><a href="../reference/blast_rec.html">blast_rec()</a></code> due to the misunderstood functionality of the <code>BLAST</code> argument (See details <a href="https://watermark.silverchair.com/bty833.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAAokwggKFBgkqhkiG9w0BBwagggJ2MIICcgIBADCCAmsGCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMvrRsd_dqJkkrO5AiAgEQgIICPNCvHwDI6PCzjXjP4FtOO2O0HPaNbVmBBOW7I0YBwK6an33k32Zm3d24U1pB5TsoPAmqK1TBBz8y9IEEynMl7IQKmupg3MXTiUqUe0oLrMDIrL4Szunm255RdiIlvq5RDR5V3B7ejVCxTbwbmvbggBHyiMCcvEEufaNROT8z7XXaUiHx3OyD1HaRkXIM4MC5x_b734rkKFIploMWjLSX5KZS1i2prd_Shn2seyV4E2cj3RVw9YzcBEZvRHpbhzWIOWSjXR3ZqYTlDCxzk5CO0d15J57u3C64sM1rQRo2Atny5KR5VL4oLxy8bBBGsQl-k6lQavisoshkBZPfLMr5kEHMrCmCtw2Q2A6pzgYC-jJ7udunpA7DK2GZsSQc-ApYXHFdEOovrtFYNHFrqlE8IsooVB7U89re0lMzZM4gYoMzcDuvLFQ1_msIl51QMmPi8yIY6E5DA2h1Ho93wUxekwTBdxB3ovAtpz4_J93w8cn4z2EK9ygHoxLRQ90kfd7nawSMTxFwBxDB0iK3_w0YtnObzcyd_JAWZM9fobNHDc1CY9j4DWKclIxhqHeCHKMtw9nB9CSa7Bccbf1EBHw4uf1-qYl30IdUsKQUpP5YOtKA7nUDTpTJiKKEf1V3MmW6rPreBayi45-QPk7GkhlrgJgaXdU4LMSGxjaDdcXhOX3W35Y95VSAZEf3rlueplgDTRnRxu8NNZ8S1c5ccA1vP0gFth80NTwzUzZy2kSITIndok_0MWSeuavaRQ7g" class="external-link">here</a> and <a href="https://github.com/drostlab/orthologr/issues/9" class="external-link">#9</a> ; Many thanks to <a href="https://github.com/armish" class="external-link">@armish</a>)</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.4" id="orthologr-version-004">
<code>orthologr</code> version 0.0.4<a class="anchor" aria-label="anchor" href="#orthologr-version-004"></a></h2>
<div class="section level3">
<h3 id="new-features-0-0-4">New Features<a class="anchor" aria-label="anchor" href="#new-features-0-0-4"></a></h3>
<ul><li>Users can now control via the new <code>delete_corrupt_cds</code> argument in <code><a href="../reference/dNdS.html">dNdS()</a></code> and related downstream functions whether or not corrupted input coding sequences shall be removed prior to dN/dS inference. In case corrupted CDS exist, the <code><a href="../reference/dNdS.html">dNdS()</a></code> function will now generate a separate fasta file which stores all corrupted CDS so that they can be investigated. See issue <a href="https://github.com/drostlab/orthologr/issues/8" class="external-link">#8</a> for details.</li>
</ul><div class="section level4">
<h4 id="function-updates-0-0-4">Function updates<a class="anchor" aria-label="anchor" href="#function-updates-0-0-4"></a></h4>
<ul><li><p><code><a href="../reference/dNdS.html">dNdS()</a></code> receives new argument <code>delete_corrupt_cds</code> to remove corrupted input coding sequences (<code>delete_corrupt_cds</code> is set to <code>TRUE</code> as default)</p></li>
<li><p><code><a href="../reference/read.cds.html">read.cds()</a></code> receives new argument <code>delete_corrupt_cds</code> to remove corrupted input coding sequences (<code>delete_corrupt_cds</code> is set to <code>TRUE</code> as default)</p></li>
<li><p><code><a href="../reference/cds2aa.html">cds2aa()</a></code> receives new argument <code>delete_corrupt_cds</code> to remove corrupted input coding sequences (<code>delete_corrupt_cds</code> is set to <code>TRUE</code> as default)</p></li>
<li><p><code><a href="../reference/set_blast.html">set_blast()</a></code> receives new argument <code>delete_corrupt_cds</code> to remove corrupted input coding sequences (<code>delete_corrupt_cds</code> is set to <code>TRUE</code> as default)</p></li>
<li><p><code><a href="../reference/blast.html">blast()</a></code> receives new argument <code>delete_corrupt_cds</code> to remove corrupted input coding sequences (<code>delete_corrupt_cds</code> is set to <code>TRUE</code> as default)</p></li>
<li><p><code><a href="../reference/blast_best.html">blast_best()</a></code> receives new argument <code>delete_corrupt_cds</code> to remove corrupted input coding sequences (<code>delete_corrupt_cds</code> is set to <code>TRUE</code> as default)</p></li>
<li><p><code><a href="../reference/blast_rec.html">blast_rec()</a></code> receives new argument <code>delete_corrupt_cds</code> to remove corrupted input coding sequences (<code>delete_corrupt_cds</code> is set to <code>TRUE</code> as default)</p></li>
</ul></div>
</div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.3" id="orthologr-version-003">
<code>orthologr</code> version 0.0.3<a class="anchor" aria-label="anchor" href="#orthologr-version-003"></a></h2>
<ul><li>Fixing internal path bug that caused that wrong pal2nal paths were generated when using multiple sequence aligners -&gt; see issue <a href="https://github.com/HajkD/orthologr/issues/5" class="external-link uri">https://github.com/HajkD/orthologr/issues/5</a> (Many thanks to Dr. Mario López-Pérez)</li>
</ul></div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.2" id="orthologr-version-002">
<code>orthologr</code> version 0.0.2<a class="anchor" aria-label="anchor" href="#orthologr-version-002"></a></h2>
<ul><li>Fixing a major bug that caused KaKs_Calculator to not be able to correctly parse the kaks computation output (Many thanks to <a href="https://github.com/lihongyi123" class="external-link">Hongyi Li</a> who spotted the bug and found a solution).</li>
</ul></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

