<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparative Genomics with R • orthologr</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="jquery.sticky-kit.min.js"></script><script src="pkgdown.js"></script><meta property="og:title" content="Comparative Genomics with R">
<meta property="og:description" content="Perform BLAST searches, pairwise and multiple sequence alignments, orthology inference, and
    population genomics for life science research.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">orthologr</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="articles/dNdS_estimation.html">dNdS estimation using orthologr</a>
    </li>
    <li>
      <a href="articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
    <div id="orthologr" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#orthologr" class="anchor"></a>orthologr</h1></div>

<div id="comparative-genomics-with-r" class="section level2">
<h2 class="hasAnchor">
<a href="#comparative-genomics-with-r" class="anchor"></a>Comparative Genomics with R</h2>
<p>The comparative method is a powerful approach in genomics research. Based on our knowledge about the phylogenetic relationships between species, we can study the evolution, diversification, and constraints of biological processes by comparing genomes, genes, and other genomic loci across species. The <code>orthologr</code> package aims to provide a framework to perform comparative genomics studies with R and implements R wrapper functions to the most important and most popular genomics and comparative genomics tools.</p>
<p>In detail, <code>orthologr</code> allows users to perform BLAST searches, orthology inference methods, multiple sequence alignments, codon alignments, dNdS estimation, and <a href="https://github.com/HajkD/myTAI/blob/master/vignettes/Introduction.Rmd">divergence stratigraphy</a> <strong>between entire genomes</strong> with R.</p>
<p>The most useful implementation in <code>orthologr</code> is the ability to compute synonymous versus non-synonymous substitution rates (dN/dS) for all orthologous genes between two <strong>entire genomes</strong>.</p>
<div id="citation" class="section level3">
<h3 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h3>
<p><strong>Please cite the following paper when using <code>orthologr</code> for your own research. This will allow me to continue working on this software tool and will motivate me to extend its functionality and usability. Many thanks in advance :)</strong></p>
<blockquote>
<p>Drost et al. 2015. <strong>Evidence for Active Maintenance of Phylotranscriptomic Hourglass Patterns in Animal and Plant Embryogenesis</strong>. <em>Mol. Biol. Evol.</em> 32 (5): 1221-1231. <a href="http://mbe.oxfordjournals.org/content/32/5/1221.abstract?sid=767aea12-1eb3-40be-8c6a-e2861f159b46">doi:10.1093/molbev/msv012</a></p>
</blockquote>
</div>
</div>
<div id="install" class="section level2">
<h2 class="hasAnchor">
<a href="#install" class="anchor"></a>Install</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># first install the devtools package </span>
<span class="kw">install.packages</span>(<span class="st">"devtools"</span>)
<span class="co"># install orthologr on your system</span>
<span class="kw">source</span>(<span class="st">"http://bioconductor.org/biocLite.R"</span>)
<span class="kw">biocLite</span>(<span class="st">"HajkD/orthologr"</span>)</code></pre></div>
</div>
<div id="use-cases" class="section level2">
<h2 class="hasAnchor">
<a href="#use-cases" class="anchor"></a>Use Cases</h2>
<p>Learn <code>orthologr</code> by reading these tutorials:</p>
<ul>
<li><a href="https://github.com/HajkD/orthologr/blob/master/vignettes/Install.Rmd">Install Prerequisite Tools</a></li>
<li><a href="https://github.com/HajkD/orthologr/blob/master/vignettes/blast.Rmd">Perform BLAST Searches</a></li>
<li><a href="https://github.com/HajkD/orthologr/blob/master/vignettes/sequence_alignments.Rmd">Perform Sequence Alignments</a></li>
<li><a href="https://github.com/HajkD/orthologr/blob/master/vignettes/orthology_inference.Rmd">Perform Orthology Inference</a></li>
<li><a href="https://github.com/HajkD/orthologr/blob/master/vignettes/dNdS_estimation.Rmd">Perform dNdS Estimation</a></li>
<li><a href="https://github.com/HajkD/orthologr/blob/master/vignettes/divergence_stratigraphy.Rmd">Perform Divergence Stratigraphy</a></li>
</ul>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<div id="small-example-with-internal-dataset" class="section level3">
<h3 class="hasAnchor">
<a href="#small-example-with-internal-dataset" class="anchor"></a>Small example with internal dataset</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(orthologr)

<span class="co"># Detect orthologous genes between a query species and a subject species</span>
<span class="co"># and compute the synonymous versus non-synonymous substitution rates (dN/dS)</span>
<span class="co"># following this paradigm:</span>
<span class="co"># 1) reciprocal best hit for orthology inference (RBH)</span>
<span class="co"># 2) Needleman-Wunsch for pairwise amino acid alignments</span>
<span class="co"># 3) pal2nal for codon alignments</span>
<span class="co"># 4) Comeron for dNdS estimation</span>
<span class="co"># 5) multi-core processing 'comp_cores = 1'</span>
<span class="kw"><a href="reference/dNdS.html">dNdS</a></span>(<span class="dt">query_file      =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
     <span class="dt">subject_file    =</span> <span class="kw">system.file</span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
     <span class="dt">ortho_detection =</span> <span class="st">"RBH"</span>, 
     <span class="dt">aa_aln_type     =</span> <span class="st">"pairwise"</span>,
     <span class="dt">aa_aln_tool     =</span> <span class="st">"NW"</span>, 
     <span class="dt">codon_aln_tool  =</span> <span class="st">"pal2nal"</span>, 
     <span class="dt">dnds_est.method =</span> <span class="st">"Comeron"</span>, 
     <span class="dt">comp_cores      =</span> <span class="dv">1</span> )</code></pre></div>
<pre><code>       query_id            subject_id       dN     dS    dNdS
 1: AT1G01010.1 333554|PACid:16033839 0.106400 0.2537 0.41950
 2: AT1G01020.1 470181|PACid:16064328 0.040230 0.1037 0.38790
 3: AT1G01030.1 470180|PACid:16054974 0.014990 0.1265 0.11850
 4: AT1G01040.1 333551|PACid:16057793 0.013470 0.1165 0.11560
 5: AT1G01050.1 909874|PACid:16064489 0.000000 0.1750 0.00000
 6: AT1G01060.3 470177|PACid:16043374 0.044950 0.1133 0.39670
 7: AT1G01070.1 918864|PACid:16052578 0.018300 0.1059 0.17280
 8: AT1G01080.1 909871|PACid:16053217 0.033980 0.1056 0.32170
 9: AT1G01090.1 470171|PACid:16052860 0.009104 0.2181 0.04174
10: AT1G01110.2 333544|PACid:16034284 0.032480 0.1220 0.26620
11: AT1G01120.1 918858|PACid:16049140 0.003072 0.1326 0.02317
12: AT1G01140.3 470161|PACid:16036015 0.005672 0.1312 0.04324
13: AT1G01150.1 918855|PACid:16037307 0.130000 0.2028 0.64120
14: AT1G01160.1 918854|PACid:16044153 0.104600 0.2804 0.37310
15: AT1G01170.2 311317|PACid:16052302 0.000000 0.3064 0.00000
16: AT1G01180.1 909860|PACid:16056125 0.029680 0.1763 0.16830
17: AT1G01190.1 311315|PACid:16059488 0.028690 0.1618 0.17730
18: AT1G01200.1 470156|PACid:16041002 0.019050 0.1675 0.11370
19: AT1G01210.1 311313|PACid:16057125 0.020670 0.1540 0.13420
20: AT1G01220.1 470155|PACid:16047984 0.015690 0.1533 0.10230</code></pre>
</div>
<div id="example-computing-dnds-values-for-all-orthologous-genes-between-two-genomes" class="section level3">
<h3 class="hasAnchor">
<a href="#example-computing-dnds-values-for-all-orthologous-genes-between-two-genomes" class="anchor"></a>Example: Computing dN/dS values for all orthologous genes between two genomes</h3>
<p>First, users can retrieve all coding sequences from entire genomes using the <a href="https://github.com/ropensci/biomartr">biomartr</a> package (<a href="https://github.com/ropensci/biomartr/blob/master/vignettes/Sequence_Retrieval.Rmd#cds-retrieval">see details here</a>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">"biomartr"</span>)
<span class="kw">library</span>(biomartr)
<span class="co"># download all coding sequences for Mus musculus</span>
Mmusculus_file &lt;-<span class="st"> </span>biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/getCDS">getCDS</a></span>(<span class="dt">organism =</span> <span class="st">"Mus musculus"</span>, <span class="dt">path =</span> <span class="kw">getwd</span>())
<span class="co"># download all coding sequences for Homo sapiens</span>
Hsapiens_file &lt;-<span class="st"> </span>biomartr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/biomartr/topics/getCDS">getCDS</a></span>(<span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>, <span class="dt">path =</span> <span class="kw">getwd</span>())
<span class="co"># compute dN/dS values for Homo sapiens versus Mus musculus</span>
Hs_vs_Mm_dNdS &lt;-<span class="st"> </span>
<span class="st">  </span><span class="kw"><a href="reference/dNdS.html">dNdS</a></span>(<span class="dt">query_file      =</span> Hsapiens_file,
       <span class="dt">subject_file    =</span> Mmusculus_file,
       <span class="dt">ortho_detection =</span> <span class="st">"RBH"</span>, 
       <span class="dt">aa_aln_type     =</span> <span class="st">"pairwise"</span>,
       <span class="dt">aa_aln_tool     =</span> <span class="st">"NW"</span>, 
       <span class="dt">codon_aln_tool  =</span> <span class="st">"pal2nal"</span>, 
       <span class="dt">dnds_est.method =</span> <span class="st">"Comeron"</span>, 
       <span class="dt">comp_cores      =</span> <span class="dv">1</span> )
<span class="co"># store result in Excel readable csv file</span>
<span class="kw">install.packages</span>(<span class="st">"readr"</span>)
readr<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/readr/topics/write_delim">write_excel_csv</a></span>(Hs_vs_Mm_dNdS, <span class="st">"Hs_vs_Mm_dNdS.csv"</span>)</code></pre></div>
<p><strong>This way, users can compute dN/dS values for any pairwise genome comparison.</strong></p>
</div>
<div id="on-windows-systems" class="section level3">
<h3 class="hasAnchor">
<a href="#on-windows-systems" class="anchor"></a>On Windows Systems</h3>
<p>In some cases (when working with <strong>WINDOWS</strong> machines), the installation via <code>devtools</code> will not work properly. In this case users can try the follwing steps:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># On Windows, this won't work - see ?build_github_devtools</span>
<span class="kw">install_github</span>(<span class="st">"HajkD/orthologr"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)

<span class="co"># When working with Windows, first users need to install the</span>
<span class="co"># R package: rtools -&gt; install.packages("rtools")</span>

<span class="co"># Afterwards users can install devtools -&gt; install.packages("devtools")</span>
<span class="co"># and then they can run:</span>

devtools<span class="op">::</span><span class="kw"><a href="http://www.rdocumentation.org/packages/devtools/topics/install_github">install_github</a></span>(<span class="st">"HajkD/orthologr"</span>, <span class="dt">build_vignettes =</span> <span class="ot">TRUE</span>, <span class="dt">dependencies =</span> <span class="ot">TRUE</span>)

<span class="co"># and then call it from the library</span>
<span class="kw">library</span>(<span class="st">"orthologr"</span>, <span class="dt">lib.loc =</span> <span class="st">"C:/Program Files/R/R-3.1.1/library"</span>)</code></pre></div>
</div>
<div id="troubleshooting-on-windows-machines" class="section level3">
<h3 class="hasAnchor">
<a href="#troubleshooting-on-windows-machines" class="anchor"></a>Troubleshooting on Windows Machines</h3>
<ul>
<li>Install <code>orthologr</code> on a Win 8 laptop: <a href="https://github.com/HajkD/orthologr/issues/1">solution</a> ( Thanks to Andres Romanowski )</li>
</ul>
</div>
</div>
<div id="interfaces-implemented-in-orthologr" class="section level2">
<h2 class="hasAnchor">
<a href="#interfaces-implemented-in-orthologr" class="anchor"></a>Interfaces implemented in <code>orthologr</code>:</h2>
<div id="perform-blast-searches-with-r" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-blast-searches-with-r" class="anchor"></a>Perform BLAST searches with R</h3>
<ul>
<li>
<code><a href="reference/advanced_blast.html">advanced_blast()</a></code>: Perform an advanced BLAST+ search</li>
<li>
<code><a href="reference/advanced_makedb.html">advanced_makedb()</a></code>: Create a BLASTable database with makeblastdb (advanced options)</li>
<li>
<code><a href="reference/blast.html">blast()</a></code>: Perform a BLAST+ search</li>
<li>
<code><a href="reference/blast.nr.html">blast.nr()</a></code>: Perform a BLASTP search against NCBI nr</li>
<li>
<code><a href="reference/blast_best.html">blast_best()</a></code>: Perform a BLAST+ best hit search</li>
<li>
<code><a href="reference/blast_rec.html">blast_rec()</a></code>: Perform a BLAST+ best reciprocal hit (BRH) search</li>
<li>
<code><a href="reference/delta.blast.html">delta.blast()</a></code>: Perform a DELTA-BLAST Search</li>
</ul>
</div>
<div id="perform-pairwise-and-multiple-sequence-alignements-with-r" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-pairwise-and-multiple-sequence-alignements-with-r" class="anchor"></a>Perform Pairwise and Multiple Sequence Alignements with R</h3>
<ul>
<li>
<code><a href="reference/multi_aln.html">multi_aln()</a></code>: Compute Multiple Sequence Alignments based on the <code>clustalw</code>, <code>t_coffee</code>, <code>muscle</code>, <code>clustalo</code>, and <code>mafft</code> programs.</li>
<li>
<code><a href="reference/pairwise_aln.html">pairwise_aln()</a></code>: Compute Pairwise Alignments</li>
<li>
<code><a href="reference/codon_aln.html">codon_aln()</a></code>: Compute a Codon Alignment</li>
</ul>
</div>
<div id="perform-orthology-inference-with-r" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-orthology-inference-with-r" class="anchor"></a>Perform Orthology Inference with R</h3>
<ul>
<li>
<code><a href="reference/orthologs.html">orthologs()</a></code>: Main Orthology Inference Function</li>
<li>
<code><a href="reference/ProteinOrtho.html">ProteinOrtho()</a></code>: Orthology Inference with ProteinOrtho</li>
</ul>
</div>
<div id="perform-population-genomics-with-r" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-population-genomics-with-r" class="anchor"></a>Perform Population Genomics with R</h3>
<ul>
<li>
<code><a href="reference/dNdS.html">dNdS()</a></code>: Compute dNdS values for two organisms</li>
<li>
<code><a href="reference/substitutionrate.html">substitutionrate()</a></code>: Internal function for dNdS computations</li>
</ul>
</div>
<div id="read-and-write-cds-genomes-and-proteomes" class="section level3">
<h3 class="hasAnchor">
<a href="#read-and-write-cds-genomes-and-proteomes" class="anchor"></a>Read and Write CDS, Genomes, and Proteomes</h3>
<ul>
<li>
<code><a href="reference/read.cds.html">read.cds()</a></code>: Read the CDS of a given organism</li>
<li>
<code><a href="reference/read.genome.html">read.genome()</a></code>: Read the genome of a given organism</li>
<li>
<code><a href="reference/read.proteome.html">read.proteome()</a></code>: Read the proteome of a given organism</li>
<li>
<code><a href="reference/write.proteome.html">write.proteome()</a></code>: Save a proteome in fasta format</li>
</ul>
</div>
</div>
<div id="discussions-and-bug-reports" class="section level2">
<h2 class="hasAnchor">
<a href="#discussions-and-bug-reports" class="anchor"></a>Discussions and Bug Reports</h2>
<p>I would be very happy to learn more about potential improvements of the concepts and functions provided in this package.</p>
<p>Furthermore, in case you find some bugs, need additional (more flexible) functionality of parts of this package, or want to contribute to this project please let me know:</p>
<p><a href="https://github.com/HajkD/orthologr/issues" class="uri">https://github.com/HajkD/orthologr/issues</a></p>
</div>
<div id="licenses" class="section level2">
<h2 class="hasAnchor">
<a href="#licenses" class="anchor"></a>Licenses</h2>
<p>The <code>orthologr</code> package includes source code that has been published under following licenses:</p>
<div id="gestimator" class="section level3">
<h3 class="hasAnchor">
<a href="#gestimator" class="anchor"></a>gestimator</h3>
<pre><code>All files included in `orthologr` that were taken from gestimator are 
also part of libsequence.

libsequence is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.

libsequence is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
long with libsequence.  If not, see &lt;http://www.gnu.org/licenses/&gt;.


Modified by Sarah Scharfenberg and Hajk-Georg Drost (2014) to work 
in orthologr without using external libraries from libsequence.

All changes are also free under the terms of GNU General Public License
version 2 of the License, or any later version.
</code></pre>
</div>
<div id="kaks_calculator" class="section level3">
<h3 class="hasAnchor">
<a href="#kaks_calculator" class="anchor"></a>KaKs_Calculator</h3>
<p>In <code>orthologr</code> the file <code>parseFastaIntoAXT.pl</code> is stored in <code>/inst/KaKs_Calc_parser</code>.</p>
<pre><code>The parseFastaIntoAXT.pl script is freely available under GNU GPL v3 
Licence and included in the KaKs_Calculator project that can be found at 
https://code.google.com/p/kaks-calculator/
</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/HajkD/orthologr">https://​github.com/​HajkD/​orthologr</a>
</li>
<li>Report a bug at <br><a href="https://github.com/HajkD/orthologr/issues">https://​github.com/​HajkD/​orthologr/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2">GPL-2</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing orthologr</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Hajk-Georg Drost <br><small class="roles"> Author, maintainer </small>  </li>
<li><a href="authors.html">All authors...</a></li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/HajkD/orthologr"><img src="https://travis-ci.org/HajkD/orthologr.svg?branch=master" alt="Travis-CI Build Status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  

  </body>
</html>
