<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>BLAST searches using orthologr • orthologr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="BLAST searches using orthologr">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">dNdS estimation using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>BLAST searches using orthologr</h1>
            
            <h4 class="date">2019-03-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/HajkD/orthologr/blob/master/vignettes/blast.Rmd"><code>vignettes/blast.Rmd</code></a></small>
      <div class="hidden name"><code>blast.Rmd</code></div>

    </div>

    
    
<ul>
<li><a href="#getting-started">1. Getting started with BLAST</a></li>
<li>
<a href="#using-orthologr-to-perform-blast-searches">2. Core BLAST+ Functionality</a>
<ul>
<li>
<a href="#the-blast-function">2.1 The blast() function</a>
<ul>
<li><a href="#working-with-one-to-one-hits">2.1.1 Inferring with one-to-one homologous hits</a></li>
<li><a href="#working-with-one-to-many-hits">2.1.2 Inferring with one-to-many homologous hits</a></li>
</ul>
</li>
<li><a href="#the-blast_best-function">2.2 The blast_best() function</a></li>
<li><a href="#the-blast_rec-function">2.3 The blast_rec() function</a></li>
<li><a href="#the-set_blast-function">2.4 The set_blast() function</a></li>
</ul>
</li>
<li><a href="#perform-bast-searches-between-genomes">3. Perform BLAST+ Searches between Genomes</a></li>
</ul>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting Started</h2>
<p>The <code>orthologr</code> package provides several interface functions to perform BLAST searches.</p>
<p>First, users need to make sure that they have <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/">BLAST</a> installed on their machine. Please follow <a href="https://github.com/HajkD/orthologr/blob/master/vignettes/Install.Rmd#install-blast">these</a> instructions to install BLAST on your manine.</p>
</div>
<div id="performing-blast-searches" class="section level2">
<h2 class="hasAnchor">
<a href="#performing-blast-searches" class="anchor"></a>Performing BLAST Searches</h2>
<p>The <code>orthologr</code> package stores 20 example genes (orthologs) between <em>Arabidopsis thaliana</em> and <em>Arabidopsis lyrata</em>. The following example BLAST search shall illustrate a simple search with standard parameters provided by the <code><a href="../reference/blast.html">blast()</a></code> function.</p>
<p>When running the subsequent functions please make sure you can call BLAST+ from your console either in the standard <code>PATH</code> or in case you have BLAST+ installed in a separate folder, please specify the <code>path</code> argument that can be passed to <code><a href="../reference/blast.html">blast()</a></code>.</p>
<p>To check whether BLAST+ can be executed from the default <code>PATH</code> (<code>usr/bin/local</code> on UNIX systems), you can run:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># check if blast is installed and if yes, what version of blast</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"blastp -version"</span>)</code></pre></div>
<p>This should return something like this:</p>
<pre><code>blastp: 2.8.1+
 Package: blast 2.8.1, build Nov 26 2018 12:45:20</code></pre>
<p>If everything works properly, you can get started with your first BLAST+ search.</p>
</div>
<div id="using-orthologr-to-perform-blast-searches" class="section level2">
<h2 class="hasAnchor">
<a href="#using-orthologr-to-perform-blast-searches" class="anchor"></a>Using orthologr to perform BLAST searches</h2>
<p>The <code>orthologr</code> packages allows users to perform fast and easy-to-use BLAST searches between <code>fasta</code> files. These searches can even be scaled towards genome and proteome comparisons.</p>
<div id="the-blast-function" class="section level3">
<h3 class="hasAnchor">
<a href="#the-blast-function" class="anchor"></a>The blast() function</h3>
<p>The <code><a href="../reference/blast.html">blast()</a></code> function provides the easiest way to perform a BLAST search.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)
<span class="co"># performing a BLAST search using blastp (default)</span>
hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast.html">blast</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))
<span class="co"># look at results</span>
<span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">glimpse</a></span>(hit_tbl)</code></pre></div>
<pre><code>Variables:
$ query_id      (chr) "AT1G01010.1", "AT1G01020.1", "AT1G01030.1",...
$ subject_id    (chr) "333554|PACid:16033839", "470181|PACid:16064...
$ perc_identity (dbl) 73.99, 91.06, 95.54, 91.98, 100.00, 89.51, 9...
$ alig_length   (dbl) 469, 246, 359, 1970, 213, 648, 366, 300, 434...
$ mismatches    (dbl) 80, 22, 12, 85, 0, 58, 14, 22, 8, 34, 4, 6, ...
$ gap_openings  (dbl) 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, 1, 3, 2, 1...
$ q_start       (dbl) 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2,...
$ q_end         (dbl) 430, 246, 359, 1910, 213, 646, 366, 294, 429...
$ s_start       (dbl) 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4...
$ s_end         (dbl) 466, 246, 355, 1963, 213, 640, 362, 299, 433...
$ evalue        (dbl) 0e+00, 7e-166, 0e+00, 0e+00, 2e-160, 0e+00, ...
$ bit_score     (dbl) 627, 454, 698, 3704, 437, 1037, 696, 491, 85...</code></pre>
<p>As you can see, the hit table shows the output of the BLAST+ search. The <code><a href="../reference/blast.html">blast()</a></code> function runs <code>blastp</code> as default BLAST+ algorithm. Different BLAST+ algorithms can be selected by specifying the <code>blast_algorithm</code> argument, e.g. <code>blast_algorithm = "tblastn"</code>. See <code><a href="../reference/blast.html">?blast</a></code> for further details. The <code><a href="../reference/blast.html">blast()</a></code> function returns the BLAST arguments: <code>query_id</code>, <code>subject_id</code>, <code>perc_identity</code>, <code>alig_length</code>, <code>mismatches</code>, <code>gap_openings</code>, <code>q_start</code>, <code>q_end</code>, <code>s_start</code>, <code>s_end</code>, <code>evalue</code>, and <code>bit_score</code>.</p>
<p>Since <code><a href="../reference/blast.html">blast()</a></code> stores the hit table returned by BLAST in a data.table object, you can access each column, using the <a href="https://github.com/Rdatatable/data.table/wiki">data.table notation</a>.</p>
<p>In case you need to specify the <code>PATH</code> to BLAST+ please use the <code>path</code> argument:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># performing a BLAST search using blastp (default)</span>
hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast.html">blast</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">path         =</span> <span class="st">"/path/to/blastp"</span>)


hit_tbl</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># access columns: query_id, subject_id, evalue, and bit_score</span>
dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(hit_tbl, query_id, subject_id, evalue, bit_score)</code></pre></div>
<pre><code>       query_id            subject_id evalue bit_score
 1: AT1G01010.1 333554|PACid:16033839  0e+00       627
 2: AT1G01020.1 470181|PACid:16064328 7e-166       454
 3: AT1G01030.1 470180|PACid:16054974  0e+00       698
 4: AT1G01040.1 333551|PACid:16057793  0e+00      3704
 5: AT1G01050.1 909874|PACid:16064489 2e-160       437
 6: AT1G01060.3 470177|PACid:16043374  0e+00      1037
 7: AT1G01070.1 918864|PACid:16052578  0e+00       696
 8: AT1G01080.1 909871|PACid:16053217 1e-178       491
 9: AT1G01090.1 470171|PACid:16052860  0e+00       859
10: AT1G01110.2 333544|PACid:16034284  0e+00       972
11: AT1G01120.1 918858|PACid:16049140  0e+00      1092
12: AT1G01140.3 470161|PACid:16036015  0e+00       918
13: AT1G01150.1 918855|PACid:16037307 3e-150       421
14: AT1G01160.1 918854|PACid:16044153  1e-93       268
15: AT1G01170.2 311317|PACid:16052302  3e-54       158
16: AT1G01180.1 909860|PACid:16056125  0e+00       576
17: AT1G01190.1 311315|PACid:16059488  0e+00      1036
18: AT1G01200.1 470156|PACid:16041002 3e-172       470
19: AT1G01210.1 311313|PACid:16057125  7e-76       215
20: AT1G01220.1 470155|PACid:16047984  0e+00      2106</code></pre>
<p>The <code><a href="../reference/blast.html">blast()</a></code> function also allows you to pass additional parameters to the BLAST+ search using the <code>blast_params</code> argument. In the following example, a remote BLAST+ search is performed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast.html">blast</a></span>(<span class="dt">query_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">blast_params =</span> <span class="st">"-qcov_hsp_perc 0.9"</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">glimpse</a></span>(hit_tbl)</code></pre></div>
<pre><code>bservations: 20
Variables: 12
$ query_id      &lt;chr&gt; "AT1G01010.1", "AT1G01020.1", "AT1G01030.1", "AT1G010…
$ subject_id    &lt;chr&gt; "333554|PACid:16033839", "470181|PACid:16064328", "47…
$ perc_identity &lt;dbl&gt; 73.987, 91.057, 95.543, 91.980, 100.000, 89.506, 95.0…
$ alig_length   &lt;dbl&gt; 469, 246, 359, 1970, 213, 648, 366, 300, 434, 528, 52…
$ mismatches    &lt;dbl&gt; 80, 22, 12, 85, 0, 58, 14, 22, 8, 34, 4, 6, 68, 19, 0…
$ gap_openings  &lt;dbl&gt; 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, 1, 3, 2, 1, 1, 1, 0…
$ q_start       &lt;dbl&gt; 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2, 1, 1, 1,…
$ q_end         &lt;dbl&gt; 430, 246, 359, 1910, 213, 646, 366, 294, 429, 528, 52…
$ s_start       &lt;dbl&gt; 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4, 1, 6, 1…
$ s_end         &lt;dbl&gt; 466, 246, 355, 1963, 213, 640, 362, 299, 433, 528, 52…
$ evalue        &lt;dbl&gt; 0.00e+00, 8.91e-168, 0.00e+00, 0.00e+00, 3.27e-162, 0…
$ bit_score     &lt;dbl&gt; 627, 454, 698, 3704, 437, 1037, 696, 491, 859, 972, 1…</code></pre>
<p>In all cases the default <code>e-value</code> BLAST+ searches is <code>1E-5</code> and the default <code>blast_algorithm</code> is <code>blastp</code>.</p>
<p>Since BLAST+ searches can be computationally expensive, it is possible to specify the <code>comp_cores</code> argument when working with an multicore machine.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># BLAST computations using the comp_cores parameter: here with 2 cores</span>
<span class="kw"><a href="../reference/blast.html">blast</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
      <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
      <span class="dt">comp_cores   =</span> <span class="dv">2</span>)</code></pre></div>
<p>The <code>query_file</code> and <code>subject_file</code> arguments specify the path to the corresponding fasta files storing the <code>CDS</code> files, <code>amino acid</code> files, or <code>genome</code> files of the query organism and subject organism of interest. Make sure that when using <code>CDS</code>files, <code>amino acid</code> files, or <code>genome</code> files the corresponding argument <code>seq_type</code> must be adapted according to the input data format.</p>
<p>Use :</p>
<ul>
<li>
<code>CDS</code> files -&gt; <code>seq_type = "cds"</code>
</li>
<li>
<code>amino acid</code> files -&gt; <code>seq_type = "protein"</code>
</li>
<li>
<code>genome</code> files -&gt; <code>seq_type = "dna"</code>
</li>
</ul>
<p>The <code>format</code> argument specifies the input file format, e.g. “fasta” or “gbk”. The <code>blast_algorithm</code> argument specifies the BLAST program (algorithm) that shall be used to perform BLAST searches, e.g. “blastp”,“blastn”,“tblastn”,etc. Again, the <code>eval</code> argument defines the default e-value that shall be chosen as best hit threshold.</p>
</div>
<div id="using-the-split-apply-combine-strategy-for-a-blast-hit-table" class="section level3">
<h3 class="hasAnchor">
<a href="#using-the-split-apply-combine-strategy-for-a-blast-hit-table" class="anchor"></a>Using the split-apply-combine strategy for a BLAST hit table</h3>
<p>All <code>blast</code> functions implemented in <code>orthologr</code> can easily be processed using the <a href="http://www.jstatsoft.org/v40/i01/paper">split-apply-combine strategy</a> to detect for example <code>one-to-one</code>, <code>one-to-many</code>, and <code>many-to-many</code> gene homology relationships.</p>
<p>Here a simple example:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(dplyr)
<span class="co"># perform a blastp search of 20 A. thaliana genes against 1000 A. lyrata genes</span>
hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast.html">blast</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                 <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))

<span class="co"># determine 'one-to-many' and 'one-to-one' gene relationships</span>
rel_hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span>(hit_tbl, query_id), <span class="dt">n_genes =</span> <span class="kw"><a href="https://dplyr.tidyverse.org/reference/n.html">n</a></span>())
<span class="co"># look at results</span>
rel_hit_tbl</code></pre></div>
<pre><code>A tibble: 20 x 2
   query_id    n_genes
   &lt;chr&gt;         &lt;int&gt;
 1 AT1G01010.1       4
 2 AT1G01020.1       1
 3 AT1G01030.1       1
 4 AT1G01040.1       1
 5 AT1G01050.1       1
 6 AT1G01060.3       2
 7 AT1G01070.1       3
 8 AT1G01080.1       5
 9 AT1G01090.1       1
10 AT1G01110.2       1
11 AT1G01120.1       3
12 AT1G01140.3      36
13 AT1G01150.1       1
14 AT1G01160.1       1
15 AT1G01170.2       1
16 AT1G01180.1       1
17 AT1G01190.1       6
18 AT1G01200.1       7
19 AT1G01210.1       1
20 AT1G01220.1       1</code></pre>
<p>Now you can sort genes into classes: <code>one-to-one</code> and <code>one-to-many</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># classify into 'one-to-one' relationships</span>
one_to_one &lt;-<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(rel_hit_tbl, n_genes <span class="op">==</span><span class="st"> </span><span class="dv">1</span>)
<span class="co"># classify into 'one-to-many' relationships</span>
one_to_many &lt;-<span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(rel_hit_tbl, n_genes <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at one_to_one</span>
one_to_one </code></pre></div>
<pre><code>A tibble: 12 x 2
   query_id    n_genes
   &lt;chr&gt;         &lt;int&gt;
 1 AT1G01020.1       1
 2 AT1G01030.1       1
 3 AT1G01040.1       1
 4 AT1G01050.1       1
 5 AT1G01090.1       1
 6 AT1G01110.2       1
 7 AT1G01150.1       1
 8 AT1G01160.1       1
 9 AT1G01170.2       1
10 AT1G01180.1       1
11 AT1G01210.1       1
12 AT1G01220.1       1</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at one_to_many</span>
one_to_many</code></pre></div>
<pre><code>A tibble: 8 x 2
  query_id    n_genes
  &lt;chr&gt;         &lt;int&gt;
1 AT1G01010.1       4
2 AT1G01060.3       2
3 AT1G01070.1       3
4 AT1G01080.1       5
5 AT1G01120.1       3
6 AT1G01140.3      36
7 AT1G01190.1       6
8 AT1G01200.1       7</code></pre>
<p>Now we can treat classes: <code>one_to_one</code> and <code>one_to_many</code> differently:</p>
</div>
<div id="working-with-one-to-one-hits" class="section level3">
<h3 class="hasAnchor">
<a href="#working-with-one-to-one-hits" class="anchor"></a>Working with one-to-one hits</h3>
<p>We can now retrieve the blast hit entries for the one-to-one hits by joining the tables <code>one_to_one</code> and <code>hit_tbl</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at the evalue, perc_identity, and alig_length of one_to_one genes</span>
oo_genes &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/join.html">inner_join</a></span>(one_to_one, hit_tbl , <span class="dt">by =</span> <span class="st">"query_id"</span>)
oo_genes</code></pre></div>
<pre><code>A tibble: 12 x 13
   query_id n_genes subject_id perc_identity alig_length mismatches
   &lt;chr&gt;      &lt;int&gt; &lt;chr&gt;              &lt;dbl&gt;       &lt;dbl&gt;      &lt;dbl&gt;
 1 AT1G010…       1 470181|PA…          91.1         246         22
 2 AT1G010…       1 470180|PA…          95.5         359         12
 3 AT1G010…       1 333551|PA…          92.0        1970         85
 4 AT1G010…       1 909874|PA…         100           213          0
 5 AT1G010…       1 470171|PA…          96.8         434          8
 6 AT1G011…       1 333544|PA…          93.6         528         34
 7 AT1G011…       1 918855|PA…          72.6         285         68
 8 AT1G011…       1 918854|PA…          84.9         179         19
 9 AT1G011…       1 311317|PA…          85.6          97          0
10 AT1G011…       1 909860|PA…          92.6         310         20
11 AT1G012…       1 311313|PA…          95.3         107          5
12 AT1G012…       1 470155|PA…          96.7        1056         35
 … with 7 more variables: gap_openings &lt;dbl&gt;, q_start &lt;dbl&gt;, q_end &lt;dbl&gt;,
   s_start &lt;dbl&gt;, s_end &lt;dbl&gt;, evalue &lt;dbl&gt;, bit_score &lt;dbl&gt;</code></pre>
<p>This way, users will have all blast information of <code>one_to_one</code> hits available.</p>
</div>
<div id="a-real-world-application-homology-inference" class="section level3">
<h3 class="hasAnchor">
<a href="#a-real-world-application-homology-inference" class="anchor"></a>A real world application: Homology Inference</h3>
<p>Using the <code>oo_genes</code> dataset above, users can use this <code>one_to_one</code> hits strategy to detect homologous genes between species.</p>
<p>For example, we could restrict <code>one_to_one</code> hits to fulfill certain (stringent) criteria to identify as a homologous hit. Here we choose the following parameter constellation to achieve this goal: <code>one_to_one</code> genes must have a minimum <code>alignment length of 300</code>, a p<code>erc_identity of &gt; 80 percent</code> and an <code>e-value &lt; 1E-5</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># filter for 'homologous' hits</span>
true_orthologs &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(oo_genes, evalue <span class="op">&lt;</span><span class="st"> </span><span class="fl">1e-5</span>, perc_identity <span class="op">&gt;</span><span class="st"> </span><span class="dv">80</span>, alig_length <span class="op">&gt;</span><span class="st"> </span><span class="dv">300</span>)
<span class="co"># look at results</span>
dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(true_orthologs, query_id, subject_id, evalue, perc_identity, alig_length)</code></pre></div>
<pre><code>      query_id            subject_id evalue perc_identity alig_length
1: AT1G01030.1 470180|PACid:16054974      0         95.54         359
2: AT1G01040.1 333551|PACid:16057793      0         91.98        1970
3: AT1G01090.1 470171|PACid:16052860      0         96.77         434
4: AT1G01110.2 333544|PACid:16034284      0         93.56         528
5: AT1G01180.1 909860|PACid:16056125      0         92.58         310
6: AT1G01220.1 470155|PACid:16047984      0         96.69        1056</code></pre>
<p>This way we could filter out a high confidence set of homologous genes from the <code>one_to_one</code> class of genes.</p>
<p>In reality most homology inference programs and methods perform way more complicated and sophisticated analyses to distinguish true orthologs from true paralogs (in-paralogs, out-paralogs, etc.). These subsequent analyses can also be performed using the above introduced split-apply-combine strategy.</p>
<p>Note, that you can perform self-BLAST searches <code><a href="../reference/blast.html">blast(query,query)</a></code> and <code><a href="../reference/blast.html">blast(subject,subject)</a></code> to distinguish between orthologous and paralogous genes.</p>
<p>Now we continue with the <code>one_to_many</code> class of genes.</p>
</div>
<div id="working-with-one-to-many-hits" class="section level3">
<h3 class="hasAnchor">
<a href="#working-with-one-to-many-hits" class="anchor"></a>Working with one-to-many hits</h3>
<p>Here we want to address the question how to deal with multiple hits returned by <code>BLAST+</code>.</p>
<p>Again we investigate all <code>one_to_many</code> genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at results </span>
one_to_many</code></pre></div>
<pre><code> A tibble: 8 x 2
  query_id    n_genes
  &lt;chr&gt;         &lt;int&gt;
1 AT1G01010.1       4
2 AT1G01060.3       2
3 AT1G01070.1       3
4 AT1G01080.1       5
5 AT1G01120.1       3
6 AT1G01140.3      36
7 AT1G01190.1       6
8 AT1G01200.1       7</code></pre>
<p>When looking at gene_id <code>AT1G01200.1</code> we see that it was found 7 times in the corresponding subject set of <em>A. lyrata</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># look at all 7 hits found</span>
dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(hit_tbl, query_id <span class="op">==</span><span class="st"> "AT1G01200.1"</span>), query_id, subject_id, evalue, perc_identity, alig_length)</code></pre></div>
<pre><code> A tibble: 7 x 5
  query_id    subject_id               evalue perc_identity alig_length
  &lt;chr&gt;       &lt;chr&gt;                     &lt;dbl&gt;         &lt;dbl&gt;       &lt;dbl&gt;
1 AT1G01200.1 470156|PACid:16041002 1.87e-172          95.8         238
2 AT1G01200.1 910431|PACid:16035207 2.63e- 75          53.0         219
3 AT1G01200.1 918732|PACid:16054958 1.07e- 51          44.6         193
4 AT1G01200.1 919287|PACid:16060536 4.42e- 70          58.1         179
5 AT1G01200.1 919355|PACid:16050170 2.40e- 73          53.3         212
6 AT1G01200.1 919721|PACid:16036935 2.63e- 81          59.3         204
7 AT1G01200.1 919852|PACid:16055066 4.02e-  7          24.0         154</code></pre>
<p>Now we have to decide which hit shall be considered as potential <em>homolog</em>.</p>
<p>In this example <code>subject_id</code> <code>470156|PACid:16041002</code> has the highest <code>perc_identity</code> as well as the lowest e-value <code>1.87e-172</code>. So a straightforward approach would be to choose subject gene <code>470156|PACid:16041002</code> as potential ortholog of query gene <code>AT1G01200.1</code>.</p>
<p>We can validate this approach by running a reciprocal best hit search with <code><a href="../reference/blast_rec.html">blast_rec()</a></code>and compare the output of gene <code>AT1G01200.1</code> with our choice <code>470156|PACid:16041002</code>.</p>
<p>A reciprocal best hit blast approach denotes the search strategy in which query sequences are blasted in one direction to detect matches in subject sequences and then inversely subject sequences are blasted in the other direction to detect matches in the query sequences. Only when both blast seaches in both directions result in the same hit pair: BLAST(A,B) = BLAST(B,A) the hit will be considered as true ortholog relationship.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># run blast best reciprocal hit function</span>
rbh_hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast_rec.html">blast_rec</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                         <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))
<span class="co"># look at results</span>
dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(rbh_hit_tbl, query_id, subject_id, evalue, perc_identity, alig_length)</code></pre></div>
<pre><code> A tibble: 20 x 5
 Groups:   query_id [20]
   query_id    subject_id               evalue perc_identity alig_length
   &lt;chr&gt;       &lt;chr&gt;                     &lt;dbl&gt;         &lt;dbl&gt;       &lt;dbl&gt;
 1 AT1G01010.1 333554|PACid:16033839 0.                 74.0         469
 2 AT1G01020.1 470181|PACid:16064328 4.33e-166          91.1         246
 3 AT1G01030.1 470180|PACid:16054974 0.                 95.5         359
 4 AT1G01040.1 333551|PACid:16057793 0.                 92.0        1970
 5 AT1G01050.1 909874|PACid:16064489 1.59e-160         100           213
 6 AT1G01060.3 470177|PACid:16043374 0.                 89.5         648
 7 AT1G01070.1 918864|PACid:16052578 0.                 95.1         366
 8 AT1G01080.1 909871|PACid:16053217 5.90e-179          90.3         300
 9 AT1G01090.1 470171|PACid:16052860 0.                 96.8         434
10 AT1G01110.2 333544|PACid:16034284 0.                 93.6         528
11 AT1G01120.1 918858|PACid:16049140 0.                 99.2         529
12 AT1G01140.3 470161|PACid:16036015 0.                 98.5         453
13 AT1G01150.1 918855|PACid:16037307 2.01e-150          72.6         285
14 AT1G01160.1 918854|PACid:16044153 1.16e- 93          84.9         179
15 AT1G01170.2 311317|PACid:16052302 5.75e- 54          85.6          97
16 AT1G01180.1 909860|PACid:16056125 0.                 92.6         310
17 AT1G01190.1 311315|PACid:16059488 0.                 94.2         533
18 AT1G01200.1 470156|PACid:16041002 1.87e-172          95.8         238
19 AT1G01210.1 311313|PACid:16057125 8.80e- 76          95.3         107
20 AT1G01220.1 470155|PACid:16047984 0.                 96.7        1056</code></pre>
<p>When we now look at gene <code>AT1G01200.1</code> we find that after an reciprocal blast approach subject gene <code>470181|PACid:16064328</code> rather than the subject gene <code>470156|PACid:16041002</code> (= result of unidirectional blast) has been detected as potential ortholog. The example illustrates the importance of using bidirectional blast to determine orthology relationships.</p>
<p>An alternative analysis that can be performed with these three candidate subject genes is the following:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># read CDS sequences of the 20 example query genes of A. thaliana</span>
Ath.cds &lt;-<span class="st"> </span>orthologr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/orthologr/topics/read.cds">read.cds</a></span>(<span class="dt">file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>), <span class="dt">format =</span> <span class="st">"fasta"</span>)
<span class="co"># read CDS sequences of the 1000 example subject genes of A. lyrata</span>
Aly.cds &lt;-<span class="st"> </span>orthologr<span class="op">::</span><span class="kw"><a href="https://www.rdocumentation.org/packages/orthologr/topics/read.cds">read.cds</a></span>(<span class="dt">file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>), <span class="dt">format =</span> <span class="st">"fasta"</span>)
<span class="co"># show the sequence of gene AT1G01070.1</span>
Ath.cds[<span class="st">"AT1G01070.1"</span> , seqs]</code></pre></div>
<pre><code>[1] "atggctggagatatgcaaggagtgagagtagtagaaaaatattcaccggtcatagtgatggtgatgtcaaatgta
gcgatgggttcggtgaatgcacttgtgaagaaagctcttgatgttggtgtgaaccatatggtcattggtgcttatcgaat
ggctatttccgctttaattttggttccctttgcctatgtcttggaaaggaaaacaagaccacaaataacgtttaggctaa
tggtcgatcatttcgtcagtggccttctcggggcgagtttgatgcagtttttctttttgcttggtctgtcgtacacgtca
gcaactgtttcgtgtgctttggtaagcatgttgcctgcaatcaccttcgctttggcccttattttcaggactgaaaatgt
gaagattctaaagaccaaagcaggaatgttgaaggtgattggaactttgatctgtataagtggagctttgttcttaacat
tttacaaaggcccacaaatatcaaactctcactctcactctcacggtggggcttcccacaacaacaacgatcaagacaag
gccaataattggcttcttggatgtctttatttaaccataggaacagtgttgctatctctatggatgttgtttcaagggac
tttaagtattaagtacccttgcaaatactcgagcacttgtcttatgtcaattttcgcggcatttcaatgtgctctcttga
gcctttacaagagcagagacgttaatgattggatcatagatgatagattcgttatcaccgtcatcatatacgctggagtg
gtaggacaagcaatgacgacggttgcaacaacatgggggattaaaaaattaggagctgtgttcgcatcggcgtttttccc
acttactctcatttcggctactctatttgatttcctcattttacacactcctttataccttggaagtgtgattggatcac
tagtgaccataacgggtctctacatgttcttgtgggggaagaacaaagaaacggaatcatcaactgcattgtcttcagga
atggataacgaagctcaatatactactcctaataaggataacgactctaagtcgcccgtttaa"</code></pre>
<p>Now users can perform a global alignment between the CDS sequences of <code>AT1G01070.1</code> and the three subject genes as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(Biostrings)
<span class="co"># perform 3 global alignments between:  AT1G01070.1 and 918864|PACid:16052578, </span>
<span class="co"># 919693|PACid:16048878, 919961|PACid:16062329</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(Aly.cds[ <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/unlist">unlist</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(hit_tbl, query_id <span class="op">==</span><span class="st"> "AT1G01070.1"</span>), subject_id)), seqs ], pairwiseAlignment, 
       <span class="dt">pattern =</span> Ath.cds[<span class="st">"AT1G01070.1"</span> , seqs], <span class="dt">type    =</span> <span class="st">"global"</span> )</code></pre></div>
<pre><code>$...
Global PairwiseAlignmentsSingleSubject (1 of 1)
pattern: [1] atggctggagatatgcaaggagtgagagta...aaggataacgactctaagtcgcccgtttaa 
subject: [1] atgggtgaaggtatgattggagtgagagta...aaggataacgactctaagtcgcccgtttaa 
score: 1768.965 

$...
Global PairwiseAlignmentsSingleSubject (1 of 1)
pattern: [1] atggctgga---gatatgcaaggagtgaga...----cgac----tctaagtcgcccgtttaa 
subject: [1] atggctaaatcagatatgc------tg---...ggttccacaaggtctatatcgcc---ttaa 
score: -2318.726 

$...
Global PairwiseAlignmentsSingleSubject (1 of 1)
pattern: [1] atggctggagatatgcaaggagtgagagta...aaggataacgactctaagtcgcccgtttaa 
subject: [1] atgagtgaggatatgggaggagtgaaagta...----------------------------aa 
score: 486.462 </code></pre>
<p><strong>Note</strong>: To obtain the score value, you need to specify the <code>scoreOnly = TRUE</code> in the <code>pairwiseAlignment</code> function.</p>
<p>As you can see, subject gene <code>918864|PACid:16052578</code> also has the highest global alignment score <code>1768.965</code> based on the Needleman-Wunsch algorithm. This strategy might help you to differentiate between border line cases.</p>
<p>The examples shown above shall demonstrate the use cases that can be performed using the <code>blast</code> functions implemented in <code>orthologr</code>.</p>
<p>Another useful analysis can be to take the length of the initial query genes into account using the <code><a href="https://www.rdocumentation.org/packages/base/topics/nchar">nchar()</a></code> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># show the length distribution of all genes</span>
<span class="co"># stored in "Ath.cds"</span>
Ath.cds[ , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nchar">nchar</a></span>(seqs)]</code></pre></div>
<pre><code>
[1] 1290  738 1077 5730  639 1938 1098  882 1287 1584 1587 1356 1038  588  252 1437 1608  714  321 3168
</code></pre>
<p>Or the length of a specific gene:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Ath.cds[<span class="st">"AT1G01070.1"</span> , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nchar">nchar</a></span>(seqs)]</code></pre></div>
<pre><code>[1] 1098</code></pre>
<p>This way you can easily visualize the length distribution of genes stored in your query organism file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Ath.cds &lt;-<span class="st"> </span><span class="kw"><a href="../reference/read.cds.html">read.cds</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds_1000.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                    <span class="dt">format =</span> <span class="st">"fasta"</span>)
                    
<span class="co"># look at sequence length distributions</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/graphics/topics/hist">hist</a></span>(Ath.cds[ , <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nchar">nchar</a></span>(seqs)], <span class="dt">breaks =</span> <span class="dv">100</span>)</code></pre></div>
</div>
</div>
<div id="the-blast_best-function" class="section level2">
<h2 class="hasAnchor">
<a href="#the-blast_best-function" class="anchor"></a>The blast_best() function</h2>
<p>For some analyses it is sufficient to perform BLAST+ best hit searches. The <code><a href="../reference/blast_best.html">blast_best()</a></code> function is optimized to perform BLAST+ best hit searches (only based on the minimum e-value) and returns the best hit when performing a BLAST+ search of a query organisms (or set of query genes) against a subject organism (or set of subject genes).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># performing gene orthology inference using the best hit (BH) method</span>
<span class="kw"><a href="../reference/blast_best.html">blast_best</a></span>(<span class="dt">query_file    =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
           <span class="dt">subject_file  =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
           <span class="dt">clean_folders =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code> A tibble: 20 x 12
 Groups:   query_id [20]
   query_id subject_id perc_identity alig_length mismatches gap_openings
   &lt;chr&gt;    &lt;chr&gt;              &lt;dbl&gt;       &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt;
 1 AT1G010… 333554|PA…          74.0         469         80            8
 2 AT1G010… 470181|PA…          91.1         246         22            0
 3 AT1G010… 470180|PA…          95.5         359         12            2
 4 AT1G010… 333551|PA…          92.0        1970         85           10
 5 AT1G010… 909874|PA…         100           213          0            0
 6 AT1G010… 470177|PA…          89.5         648         58            5
 7 AT1G010… 918864|PA…          95.1         366         14            2
 8 AT1G010… 909871|PA…          90.3         300         22            2
 9 AT1G010… 470171|PA…          96.8         434          8            3
10 AT1G011… 333544|PA…          93.6         528         34            0
11 AT1G011… 918858|PA…          99.2         529          4            0
12 AT1G011… 470161|PA…          98.5         453          6            1
13 AT1G011… 918855|PA…          72.6         285         68            3
14 AT1G011… 918854|PA…          84.9         179         19            2
15 AT1G011… 311317|PA…          85.6          97          0            1
16 AT1G011… 909860|PA…          92.6         310         20            1
17 AT1G011… 311315|PA…          94.2         533         30            1
18 AT1G012… 470156|PA…          95.8         238         10            0
19 AT1G012… 311313|PA…          95.3         107          5            0
20 AT1G012… 470155|PA…          96.7        1056         35            0
 … with 6 more variables: q_start &lt;dbl&gt;, q_end &lt;dbl&gt;, s_start &lt;dbl&gt;,
   s_end &lt;dbl&gt;, evalue &lt;dbl&gt;, bit_score &lt;dbl&gt;</code></pre>
<p>The <code><a href="../reference/blast_best.html">blast_best()</a></code> function returns: <code>query_id</code>, <code>subject_id</code>, and <code>eval</code>.</p>
<p>In case you need more parameters returned by a BLAST+ best hit search, you can specify the <code>detailed_output</code> argument (<code>detailed_output = TRUE</code>).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># BLAST+ best hit search</span>
best_hit_tbl &lt;-<span class="st"> </span><span class="kw"><a href="../reference/blast_best.html">blast_best</a></span>(<span class="dt">query_file      =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
                           <span class="dt">subject_file    =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))
<span class="co"># look at results</span>
dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/reexports.html">glimpse</a></span>(best_hit_tbl)</code></pre></div>
<pre><code>Observations: 20
Variables: 12
Groups: query_id [20]
$ query_id      &lt;chr&gt; "AT1G01010.1", "AT1G01020.1", "AT1G01030.1", "AT1G010…
$ subject_id    &lt;chr&gt; "333554|PACid:16033839", "470181|PACid:16064328", "47…
$ perc_identity &lt;dbl&gt; 73.987, 91.057, 95.543, 91.980, 100.000, 89.506, 95.0…
$ alig_length   &lt;dbl&gt; 469, 246, 359, 1970, 213, 648, 366, 300, 434, 528, 52…
$ mismatches    &lt;dbl&gt; 80, 22, 12, 85, 0, 58, 14, 22, 8, 34, 4, 6, 68, 19, 0…
$ gap_openings  &lt;dbl&gt; 8, 0, 2, 10, 0, 5, 2, 2, 3, 0, 0, 1, 3, 2, 1, 1, 1, 0…
$ q_start       &lt;dbl&gt; 1, 1, 1, 6, 1, 1, 1, 1, 1, 1, 1, 1, 5, 4, 2, 1, 1, 1,…
$ q_end         &lt;dbl&gt; 430, 246, 359, 1910, 213, 646, 366, 294, 429, 528, 52…
$ s_start       &lt;dbl&gt; 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 16, 2, 4, 1, 6, 1…
$ s_end         &lt;dbl&gt; 466, 246, 355, 1963, 213, 640, 362, 299, 433, 528, 52…
$ evalue        &lt;dbl&gt; 0.00e+00, 8.91e-168, 0.00e+00, 0.00e+00, 3.27e-162, 0…
$ bit_score     &lt;dbl&gt; 627, 454, 698, 3704, 437, 1037, 696, 491, 859, 972, 1…</code></pre>
</div>
<div id="the-blast_rec-function" class="section level2">
<h2 class="hasAnchor">
<a href="#the-blast_rec-function" class="anchor"></a>The blast_rec() function</h2>
<p>The <code><a href="../reference/blast_rec.html">blast_rec()</a></code> function was implemented to optimize BLAST+ reciprocal best hit searches (only based on the minimum e-value). BLAST+ reciprocal best hit searches are used to perform orthology inference.</p>
<p>Running <code><a href="../reference/blast_rec.html">blast_rec()</a></code> using default parameter settings:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># performing gene orthology inference using the reciprocal best hit (RBH) method</span>
<span class="kw"><a href="../reference/blast_rec.html">blast_rec</a></span>(<span class="dt">query_file   =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),
          <span class="dt">subject_file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_lyra_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))</code></pre></div>
<pre><code> A tibble: 20 x 12
 Groups:   query_id [?]
   query_id subject_id perc_identity alig_length mismatches gap_openings
   &lt;chr&gt;    &lt;chr&gt;              &lt;dbl&gt;       &lt;dbl&gt;      &lt;dbl&gt;        &lt;dbl&gt;
 1 AT1G010… 333554|PA…          74.0         469         80            8
 2 AT1G010… 470181|PA…          91.1         246         22            0
 3 AT1G010… 470180|PA…          95.5         359         12            2
 4 AT1G010… 333551|PA…          92.0        1970         85           10
 5 AT1G010… 909874|PA…         100           213          0            0
 6 AT1G010… 470177|PA…          89.5         648         58            5
 7 AT1G010… 918864|PA…          95.1         366         14            2
 8 AT1G010… 909871|PA…          90.3         300         22            2
 9 AT1G010… 470171|PA…          96.8         434          8            3
10 AT1G011… 333544|PA…          93.6         528         34            0
11 AT1G011… 918858|PA…          99.2         529          4            0
12 AT1G011… 470161|PA…          98.5         453          6            1
13 AT1G011… 918855|PA…          72.6         285         68            3
14 AT1G011… 918854|PA…          84.9         179         19            2
15 AT1G011… 311317|PA…          85.6          97          0            1
16 AT1G011… 909860|PA…          92.6         310         20            1
17 AT1G011… 311315|PA…          94.2         533         30            1
18 AT1G012… 470156|PA…          95.8         238         10            0
19 AT1G012… 311313|PA…          95.3         107          5            0
20 AT1G012… 470155|PA…          96.7        1056         35            0
 … with 6 more variables: q_start &lt;dbl&gt;, q_end &lt;dbl&gt;, s_start &lt;dbl&gt;,
   s_end &lt;dbl&gt;, evalue &lt;dbl&gt;, bit_score &lt;dbl&gt;</code></pre>
<div id="the-set_blast-function" class="section level3">
<h3 class="hasAnchor">
<a href="#the-set_blast-function" class="anchor"></a>The set_blast() function</h3>
<p>The <code><a href="../reference/set_blast.html">set_blast()</a></code>function reads a file storing a specific sequence type, such as “cds”, “protein”, or “dna” in a standard sequence file format such as “fasta”, etc. and depending of the makedb parameter either creates a blast-able database, or returns the corresponding protein sequences as data.table object for further BLAST searches.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># using set_blast() to generate a blastable sequence database</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(<span class="kw"><a href="../reference/set_blast.html">set_blast</a></span>(<span class="dt">file =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/ortho_thal_cds.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>))[[<span class="dv">1</span>]] , <span class="dv">2</span>)</code></pre></div>
<pre><code>
       geneids
1: AT1G01010.1
2: AT1G01020.1

seqs
1: atggaggatcaagttgggtttgggttccgtccgaacgacgaggagctcgttggtcactatctccgtaacaaaatcgaaggaaacact
agccgcgacgttgaagtagccatcagcgaggtcaacatctgtagctacgatccttggaacttgcgcttccagtcaaagtacaaatcgaga
gatgctatgtggtacttcttctctcgtagagaaaacaacaaagggaatcgacagagcaggacaacggtttctggtaaatggaagcttacc
ggagaatctgttgaggtcaaggaccagtggggattttgtagtgagggctttcgtggtaagattggtcataaaagggttttggtgttcctc
gatggaagataccctgacaaaaccaaatctgattgggttatccacgagttccactacgacctcttaccagaacatcagaggacatatgtc
atctgcagacttgagtacaagggtgatgatgcggacattctatctgcttatgcaatagatcccactcccgcttttgtccccaatatgact
agtagtgcaggttctgtggtcaaccaatcacgtcaacgaaattcaggatcttacaacacttactctgagtatgattcagcaaatcatggc
cagcagtttaatgaaaactctaacattatgcagcagcaaccacttcaaggatcattcaaccctctccttgagtatgattttgcaaatcac
ggcggtcagtggctgagtgactatatcgacctgcaacagcaagttccttacttggcaccttatgaaaatgagtcggagatgatttggaag
catgtgattgaagaaaattttgagtttttggtagatgaaaggacatctatgcaacagcattacagtgatcaccggcccaaaaaacctgtg
tctggggttttgcctgatgatagcagtgatactgaaactggatcaatgattttcgaagacacttcgagctccactgatagtgttggtagt
tcagatgaaccgggccatactcgtatagatgatattccatcattgaacattattgagcctttgcacaattataaggcacaagagcaacca
aagcagcagagcaaagaaaaggtgataagttcgcagaaaagcgaatgcgagtggaaaatggctgaagactcgatcaagatacctccatcc
accaacacggtgaagcagagctggattgttttggagaatgcacagtggaactatctcaagaacatgatcattggtgtcttgttgttcatc
tccgtcattagttggatcattcttgttggttaa
2:
atggcggcgagtgaacacagatgcgtgggatgtggttttagggtaaagtcattgttcattcaatactctccgggtaacattcgtctcatg
aaatgcggaaattgcaaggaagtagcagatgagtacatcgagtgtgaacgcatgattattttcatcgatttaatccttcacagaccaaag
gtatatagacacgtcctctacaatgcaattaatccagcaactgtcaatattcagcatctgttgtggaagttggtcttcgcctatcttctt
ctagactgttatagaagcttgctactgagaaaaagtgatgaagaatcgagcttttctgatagccctgttcttctatctataaaggttctg
attggtgtcttatctgcaaacgctgcatttatcatctcttttgccattgcgactaagggtttgctaaatgaagtttccagaagaagagag
attatgttggggatattcatctctagttacttcaagatatttctgcttgcgatgttggtatgggaattcccaatgtcagtgatttttttt
gtcgatatacttctcttaacatcaaactccatggctcttaaagtgatgactgaatcaacaatgaccagatgcatagccgtatgcttaatc
gcgcacttgattagattcttggtgggtcagatttttgagccgacaatatttttgatacaaattggatctctgttgcaatatatgtcttat
tttttcagaatcgtatga

aa
1: MEDQVGFGFRPNDEELVGHYLRNKIEGNTSRDVEVAISEVNICSYDPWNLRFQSKYKSRDAMWYFFSRRENNKGNRQSRTTVSGKWK
LTGESVEVKDQWGFCSEGFRGKIGHKRVLVFLDGRYPDKTKSDWVIHEFHYDLLPEHQRTYVICRLEYKGDDADILSAYAIDPTPAFVPN
MTSSAGSVVNQSRQRNSGSYNTYSEYDSANHGQQFNENSNIMQQQPLQGSFNPLLEYDFANHGGQWLSDYIDLQQQVPYLAPYENESEMI
WKHVIEENFEFLVDERTSMQQHYSDHRPKKPVSGVLPDDSSDTETGSMIFEDTSSSTDSVGSSDEPGHTRIDDIPSLNIIEPLHNYKAQE
QPKQQSKEKVISSQKSECEWKMAEDSIKIPPSTNTVKQSWIVLENAQWNYLKNMIIGVLLFISVISWIILVG*
2:
MAASEHRCVGCGFRVKSLFIQYSPGNIRLMKCGNCKEVADEYIECERMIIFIDLILHRPKVYRHVLYNAINPATVNIQHLLWKLVFAYLL
LDCYRSLLLRKSDEESSFSDSPVLLSIKVLIGVLSANAAFIISFAIATKGLLNEVSRRREIMLGIFISSYFKIFLLAMLVWEFPMSVIFF
VDILLLTSNSMALKVMTESTMTRCIAVCLIAHLIRFLVGQIFEPTIFLIQIGSLLQYMSYFFRIV*</code></pre>
</div>
</div>
<div id="perform-blast-searches-between-genomes" class="section level2">
<h2 class="hasAnchor">
<a href="#perform-blast-searches-between-genomes" class="anchor"></a>Perform BLAST Searches between Genomes</h2>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#getting-started">Getting Started</a></li>
      <li><a href="#performing-blast-searches">Performing BLAST Searches</a></li>
      <li><a href="#using-orthologr-to-perform-blast-searches">Using orthologr to perform BLAST searches</a></li>
      <li><a href="#the-blast_best-function">The blast_best() function</a></li>
      <li><a href="#the-blast_rec-function">The blast_rec() function</a></li>
      <li><a href="#perform-blast-searches-between-genomes">Perform BLAST Searches between Genomes</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
