<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Installing Prerequisite Tools • orthologr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Installing Prerequisite Tools">
<meta property="og:description" content="orthologr">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.4.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">Orthology inference dNdS estimation of orthologous genes using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/drostlab/orthologr/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Installing Prerequisite Tools</h1>
            
            <h4 data-toc-skip class="date">2023-09-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/drostlab/orthologr/blob/HEAD/vignettes/Install.Rmd" class="external-link"><code>vignettes/Install.Rmd</code></a></small>
      <div class="hidden name"><code>Install.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="installing-prerequisite-tools">Installing Prerequisite Tools<a class="anchor" aria-label="anchor" href="#installing-prerequisite-tools"></a>
</h2>
<p>Most <code>orthologr</code> functions are interface functions that
pass data to common bioinformatics tools, internally call the
corresponding tool, and read their output as R object. For this purpose,
when using interface functions in <code>orthologr</code> users need to
install the underlying bioinformatics tools to obain accurate
results.</p>
<p>The following sections provide step by step instructions or guidance
on installing all bioinformatics tools for which R interface functions
are implemented in <code>orthologr</code>.</p>
<p><strong>Some tools are not trivial to install, so please read the
corresponding sections carefully and execute test cases that are
presented in each section.</strong></p>
</div>
<div class="section level2">
<h2 id="programming-languages">Programming Languages<a class="anchor" aria-label="anchor" href="#programming-languages"></a>
</h2>
<p>The following bioinformatics tools you are going to install are based
on the these programming languages:</p>
<ul>
<li><p><a href="http://www.cran.r-project.org" class="external-link"><strong>R</strong></a>
&gt;= 3.1.1</p></li>
<li><p><a href="http://isocpp.org/about" class="external-link"><strong>C++11</strong></a></p></li>
<li><p><a href="https://www.perl.org" class="external-link"><strong>Perl</strong></a> &gt;=
5.12</p></li>
</ul>
<p>Please make sure these programming languages are installed and
executable on the machines you are going to run <code>orthologr</code>
on.</p>
<div class="section level3">
<h3 id="install-blast">Install <code>BLAST</code><a class="anchor" aria-label="anchor" href="#install-blast"></a>
</h3>
<p><a href="http://www.ncbi.nlm.nih.gov/guide/howto/run-blast-local/%5D" class="external-link"><strong>BLAST</strong></a>
= Basic Local Alignment Search Tool finds regions of similarity between
biological sequences and is also used as underlying paradigm of most
fast orthology inference methods.</p>
<ol style="list-style-type: decimal">
<li><p>Go to <a href="ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/" class="uri">ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/</a>
and download the system specific BLAST program.</p></li>
<li><p>Install BLAST :</p></li>
</ol>
<ul>
<li>On a Windows machine (see <a href="http://www.ncbi.nlm.nih.gov/books/NBK52637/" class="external-link">installation manual:
Windows</a>) -&gt; Please carefully read the
<code>Environment Variables</code> section of the
<code>installation manual: Windows</code> and make sure the execution
<code>PATH</code> variable is set correctly.</li>
<li>On a Unix machine (see <a href="http://www.ncbi.nlm.nih.gov/books/NBK52640/" class="external-link">installation manual:
Unix</a>) -&gt; Please carefully read the <code>Configuration</code>
section of the <code>installation manual: Unix</code> and make sure the
execution <code>PATH</code> variable is set correctly to
<code>usr/local/bin</code>.</li>
</ul>
<p>For example for Linux systems open the <code>Terminal</code>
application and run (Thanks to Alexander Gabel):</p>
<pre><code># download BLAST+ version 2.2.31
wget ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/2.2.31/ncbi-blast-2.2.31+-x64-linux.tar.gz

# extract the compiled version of BLAST
tar zxvpf ncbi-blast+2.2.31+-x64-linux.tar.gz

# copy BLAST files to `usr/local/bin`
cp ncbi-blast-2.2.31+/bin/* usr/local/bin</code></pre>
<p>Alternatively users can set the system call to the BLAST programs by
specifying the <code>PATH</code> variable (this is useful, because it
allows an easier update of BLAST versions instead of deleting all BLAST
programs from <code>usr/local/bin</code>):</p>
<pre><code># open vim text editor
vi .bash_profile

# type 'Shift' then I to edit the file .bash_profile
# and specify the export PATH
export PATH=${PATH}:/path/to/downloaded/blast/folder/ncbi-blast-2.2.31+/bin

# type 'ESC' then ':' then 'w' then 'q' to save and quit the .bash_profile file 

# log out from your server with
exit

# log in again and type
blastp -version</code></pre>
<p>Now users should see the BLAST command line options.</p>
</div>
<div class="section level3">
<h3 id="some-tips">Some tips<a class="anchor" aria-label="anchor" href="#some-tips"></a>
</h3>
<p>Based on our personal experience the installation of BLAST works best
when copy/pasting the BLAST executables to the path
<code>usr/local/bin</code>. In detail you can run the following steps to
copy/paste the BLAST executables to <code>usr/local/bin</code> (on Unix
systems). However, updating BLAST will then need to manually delete all
previous BLAST programs from <code>usr/local/bin</code> :</p>
<p>Open the Terminal application on your system and type:</p>
<pre><code><span><span class="va">open</span> <span class="op">/</span><span class="va">usr</span><span class="op">/</span><span class="va">local</span><span class="op">/</span><span class="va">bin</span> </span></code></pre>
<p>Next, copy/paste the <code>blastp</code>, <code>makeblastdb</code>,
etc files (BLAST executables) from your BLAST folder to
<code>/usr/local/bin</code>. To do so you will need to enter the system
password to allow the copy process.</p>
<p>After installing the BLAST program you can open an R session and type
the following command to check whether or not BLAST can be executed from
R.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># test whether blastp is correctly installed on your machine</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"blastp -version"</span><span class="op">)</span></span></code></pre></div>
<pre><code>blastp: 2.2.31+
Package: blast 2.2.31, build Oct 27 2014 17:10:51</code></pre>
<p><strong>You should see this output if BLAST was installed
correctly.</strong></p>
<p>In case you find the following output:</p>
<pre><code>sh: blastp: command not found</code></pre>
<p>You should return to <code>step 2)</code> and install BLAST so that
it can be executed from the default execution <code>PATH</code>.</p>
<p>These interface functions to BLAST+ are implemented in
<code>orthologr</code>:</p>
<ul>
<li>
<code><a href="../reference/blast.html">blast()</a></code> : Interface function to BLAST+</li>
<li>
<code><a href="../reference/blast_best.html">blast_best()</a></code> : Perform a BLAST+ best hit search</li>
<li>
<code><a href="../reference/blast_rec.html">blast_rec()</a></code> : Perform a BLAST+ reciprocal best hit
(RBH) search</li>
<li>
<code><a href="../reference/set_blast.html">set_blast()</a></code> : Preparing the parameters and databases
for subsequent BLAST+ searches</li>
<li>
<code>blast.nr()</code> : Perform a BLASTp search against NCBI
nr</li>
<li>
<code>delta.blast()</code> : Perform a DELTA-BLAST Search</li>
<li>
<code>advanced_blast()</code> : Advanced interface function to
BLAST+</li>
<li>
<code>advanced_makedb()</code> : Advanced interface function to
makeblastdb</li>
</ul>
</div>
<div class="section level3">
<h3 id="multiple-sequence-alignment-tools">Multiple Sequence Alignment Tools<a class="anchor" aria-label="anchor" href="#multiple-sequence-alignment-tools"></a>
</h3>
<p>The <code>orthologr</code> package also provides interfaces to the
following Multiple Alignment Tools. Nevertheless, non of them have to be
installed if the corresponding interface functions are not used.</p>
</div>
<div class="section level3">
<h3 id="install-clustalw2">Install <code>ClustalW2</code><a class="anchor" aria-label="anchor" href="#install-clustalw2"></a>
</h3>
<p>To install <code>ClustalW2</code> please go to the <a href="http://www.clustal.org/clustal2/" class="external-link">ClustalW</a> homepage and
download the corresponding <a href="http://www.clustal.org/download/current/" class="external-link">clustalw2 program</a>
matching your operating system.</p>
<p>After downloading and unpacking the <code>clustalw2</code> program,
please go to the clustalw-2.1 folder and open a <code>Terminal</code>
application to type (in this example for Mac OS X):</p>
<pre><code># copy clustalw2 files to `usr/local/bin`
cp clustalw2 usr/local/bin</code></pre>
</div>
<div class="section level3">
<h3 id="install-t-coffee">Install <code>T-Coffee</code><a class="anchor" aria-label="anchor" href="#install-t-coffee"></a>
</h3>
<p>To install <code>T-Coffee</code> please go to the <a href="http://www.tcoffee.org/Projects/tcoffee/" class="external-link">T-Coffee</a> homepage
and download the corresponding <a href="http://www.tcoffee.org/Packages/Stable/Latest/" class="external-link">T-Coffee
program</a> matching your operating system.</p>
</div>
<div class="section level3">
<h3 id="install-muscle">Install <code>MUSCLE</code><a class="anchor" aria-label="anchor" href="#install-muscle"></a>
</h3>
<ul>
<li>
<a href="http://www.drive5.com/muscle/" class="external-link"><strong>MUSCLE</strong></a>
: Fast and accurate multiple alignment tool of nucleic acid and protein
sequences</li>
</ul>
</div>
<div class="section level3">
<h3 id="install-clustalo">Install <code>ClustalO</code><a class="anchor" aria-label="anchor" href="#install-clustalo"></a>
</h3>
<ol style="list-style-type: decimal">
<li><p>Download the <a href="http://argtable.sourceforge.net/" class="external-link">argtable</a> program.</p></li>
<li><p>Unzip the file.</p></li>
<li><p>Run within the argtable folder:</p></li>
</ol>
<pre><code>./configure

make

make check

sudo make install</code></pre>
<ol start="4" style="list-style-type: decimal">
<li><p>Download <a href="http://www.clustal.org/omega/" class="external-link">ClustalO</a>.</p></li>
<li><p>Unzip the folder and run within the folder:</p></li>
</ol>
<pre><code>./configure

make

sudo make install</code></pre>
</div>
<div class="section level3">
<h3 id="install-mafft">Install <code>MAFFT</code><a class="anchor" aria-label="anchor" href="#install-mafft"></a>
</h3>
<ul>
<li>
<a href="http://mafft.cbrc.jp/alignment/software/" class="external-link"><strong>MAFFT</strong></a>
: A tool for multiple sequence alignment and phylogeny</li>
</ul>
<p>In <code>orthologr</code> the function <code><a href="../reference/multi_aln.html">multi_aln()</a></code>
provides interfaces to all of these multiple alignment tools as well as
an pairwise alignment interface to the <a href="http://www.bioconductor.org/packages/release/bioc/html/Biostrings.html" class="external-link">Biostrings</a>
package performing a <a href="http://www.sciencedirect.com/science/article/pii/0022283670900574" class="external-link">Needleman-Wunsch
algorithm</a>.</p>
</div>
<div class="section level3">
<h3 id="codon-alignment-tools">Codon Alignment Tools<a class="anchor" aria-label="anchor" href="#codon-alignment-tools"></a>
</h3>
<p>The codon alignment tool <a href="http://www.bork.embl.de/pal2nal/" class="external-link">Pal2Nal</a> is already
integrated in the <code>orthologr</code> package and doesn’t need to be
installed.</p>
<ul>
<li><a href="http://www.bork.embl.de/pal2nal/" class="external-link"><strong>Pal2Nal</strong></a></li>
</ul>
<p>You don’t need to worry about downloading and installing
<strong>PAL2NAL</strong>, it is already included in the
<code>orthologr</code> package. The corresponding function
<code><a href="../reference/codon_aln.html">codon_aln()</a></code> takes a protein alignment and the corresponding
coding sequences and returns a codon alignment by calling <a href="http://www.bork.embl.de/pal2nal/" class="external-link"><strong>Pal2Nal</strong></a>
from inside of the <code>orthologr</code> package.</p>
</div>
<div class="section level3">
<h3 id="dnds-estimation-methods">dNdS Estimation Methods<a class="anchor" aria-label="anchor" href="#dnds-estimation-methods"></a>
</h3>
<p>dNdS estimation is a method to quantify the selection pressure acting
on a specific protein sequence determined by pairwise comparisons of
amino acid substitutions between two protein sequences and their
corresponding codon alignments. Different models have been proposed to
estimate this ratio quantifying selection pressure on proteins. The
<code>orthologr</code> package includes the most common dNdS estimation
methods.</p>
<p>Starting with an codon alignment returned by <code><a href="../reference/codon_aln.html">codon_aln()</a></code>
the function <code><a href="../reference/dNdS.html">dNdS()</a></code> computes the the dN, dS, and dNdS
values of pairs of proteins.</p>
<p>Based on implementations provided by <code>gestimator</code>,
<code>ape</code>, and <a href="https://code.google.com/p/kaks-calculator/" class="external-link">KaKs_Calculator</a>,
the following dNdS Estimation Methods are available in
<code>orthologr</code>:</p>
<ul>
<li><p><a href="http://link.springer.com/article/10.1007/BF02407308#page-1" class="external-link">Li</a>
: Li’s method (1993) -&gt; provided by the
<code>ape package</code></p></li>
<li><p><a href="http://link.springer.com/article/10.1007/BF00173196" class="external-link">Comeron</a> :
Comeron’s method (1995)</p></li>
<li><p><a href="http://mbe.oxfordjournals.org/content/3/5/418.short" class="external-link">NG</a> : Nei,
M. and Gojobori, T. (1986)</p></li>
<li><p><a href="http://mbe.oxfordjournals.org/content/2/2/150.short" class="external-link">LWL</a> : Li,
W.H., et al. (1985)</p></li>
<li><p><a href="http://mbe.oxfordjournals.org/content/21/12/2290.short" class="external-link">MLWL</a>
(Modified LWL), MLPB (Modified LPB): Tzeng, Y.H., et al. (2004)</p></li>
<li><p><a href="http://mbe.oxfordjournals.org/content/17/1/32.short" class="external-link">YN</a> :
Yang, Z. and Nielsen, R. (2000)</p></li>
<li><p><a href="http://www.biomedcentral.com/1471-2148/6/44/" class="external-link">MYN</a>
(Modified YN): Zhang, Z., et al. (2006)</p></li>
</ul>
<p>For this purpose you need to have <strong>KaKs_Calculator</strong>
installed on your system and executable from your default
<code>PATH</code>, e,g, <code>/usr/local/bin/</code>.</p>
</div>
<div class="section level3">
<h3 id="install-kaks_calculator-for-linuxunixos">Install KaKs_Calculator (For Linux/Unix/OS)<a class="anchor" aria-label="anchor" href="#install-kaks_calculator-for-linuxunixos"></a>
</h3>
<p>Please go to the <a href="https://code.google.com/archive/p/kaks-calculator/downloads" class="external-link">KaKs_Calculator
homepage</a> and download KaKs_Calculator.</p>
<p>E.g.</p>
<pre class="shell"><code># download KaKs_Calculator
wget https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/kaks-calculator/KaKs_Calculator1.2.tar.gz
# unzip
gzip -d KaKs_Calculator1.2.tar.gz
tar -xf KaKs_Calculator1.2.tar
# install
cd KaKs_Calculator1.2/src
sudo make
sudo cp KaKs_Calculator /usr/local/bin/</code></pre>
<p>Now you should be able to run KaKs_Calculator via
<code>KaKs_Calculator -h</code> in your bash or as
<code>system("KaKs_Calculator -h")</code> in R.</p>
<p>The most recent version <code>KaKs_Calculator2.0</code> can be found
<a href="https://sourceforge.net/projects/kakscalculator2/" class="external-link">here</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
