<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Sequence alignments with orthologr â€¢ orthologr</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Sequence alignments with orthologr">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">orthologr</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Install.html">Installing Prerequisite Tools</a>
    </li>
    <li>
      <a href="../articles/blast.html">BLAST searches using orthologr</a>
    </li>
    <li>
      <a href="../articles/dNdS_estimation.html">dNdS estimation using orthologr</a>
    </li>
    <li>
      <a href="../articles/divergence_stratigraphy.html">Performing Divergence Stratigraphy</a>
    </li>
    <li>
      <a href="../articles/orthology_inference.html">Orthology Inference using orthologr</a>
    </li>
    <li>
      <a href="../articles/sequence_alignments.html">Sequence alignments with orthologr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/HajkD/orthologr">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Sequence alignments with orthologr</h1>
            
            <h4 class="date">2019-08-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/HajkD/orthologr/blob/master/vignettes/sequence_alignments.Rmd"><code>vignettes/sequence_alignments.Rmd</code></a></small>
      <div class="hidden name"><code>sequence_alignments.Rmd</code></div>

    </div>

    
    
<p>The <code>orthologr</code> package provides multiple functions to perform pairwise and multiple sequence alignments. The following functions are implemented in <code>orthologr</code>:</p>
<ul>
<li>
<code><a href="../reference/multi_aln.html">multi_aln()</a></code> : Perform Multiple Sequence Alignments</li>
<li>
<code><a href="../reference/pairwise_aln.html">pairwise_aln()</a></code> : Perform Pairwise Sequence Alignments</li>
<li>
<code><a href="../reference/codon_aln.html">codon_aln()</a></code> : Perform Codon Alignments</li>
</ul>
<div id="getting-started" class="section level2">
<h2 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting Started</h2>
<p>Prior to be able to use all sequence alignment functions implemented in <code>orthologr</code> you need to install corresponding alignment tools of interest. The above mentioned functions provide interfaces to the following alignment programs:</p>
<div id="the-multi_aln-function" class="section level3">
<h3 class="hasAnchor">
<a href="#the-multi_aln-function" class="anchor"></a>The <code><a href="../reference/multi_aln.html">multi_aln()</a></code> function</h3>
<ul>
<li><p><a href="http://www.clustal.org/clustal2/"><strong>ClustalW</strong></a> : Advanced multiple alignment tool of nucleic acid and protein sequences</p></li>
<li><p><a href="http://www.tcoffee.org/Projects/tcoffee/"><strong>T_Coffee</strong></a> : A collection of tools for processing multiple sequence alignments of nucleic acids and proteins as well as their 3D structure</p></li>
<li><p><a href="http://www.drive5.com/muscle/"><strong>MUSCLE</strong></a> : Fast and accurate multiple alignment tool of nucleic acid and protein sequences</p></li>
<li><p><a href="http://www.clustal.org/omega/"><strong>ClustalO</strong></a> : Fast and scalable multiple alignment tool for nucleic acid and protein sequences that is also capable of performing HMM alignments</p></li>
<li><p><a href="http://mafft.cbrc.jp/alignment/software/"><strong>MAFFT</strong></a> : A tool for multiple sequence alignment and phylogeny</p></li>
</ul>
<p>The easiest way to use the <code><a href="../reference/multi_aln.html">multi_aln()</a></code> function is to store the corresponding multiple sequence alignment tools in the default execution <code>PATH</code> of you system (e.g. <code>/usr/local/bin</code> on UNIX machines).</p>
<p>You can test whether the corresponding multiple sequence alignment tool can be executed from the default <code>PATH</code> by running:</p>
</div>
<div id="clustalw" class="section level3">
<h3 class="hasAnchor">
<a href="#clustalw" class="anchor"></a>ClustalW</h3>
<ul>
<li><p>MacOS: <code><a href="https://www.rdocumentation.org/packages/base/topics/system">system("clustalw2 -help")</a></code></p></li>
<li><p>Linux: <code><a href="https://www.rdocumentation.org/packages/base/topics/system">system("clustalw -help")</a></code></p></li>
<li><p>Windows: <code><a href="https://www.rdocumentation.org/packages/base/topics/system">system("clustalw2.exe -help")</a></code></p></li>
</ul>
<p>In case everything is installed appropriately, you should see:</p>
<pre><code>CLUSTAL 2.1 Multiple Sequence Alignments


                DATA (sequences)

-INFILE=file.ext                             :input sequences.
-PROFILE1=file.ext  and  -PROFILE2=file.ext  :profiles (old alignment).


                VERBS (do things)

-OPTIONS            :list the command line parameters
-HELP  or -CHECK    :outline the command line params.
-FULLHELP           :output full help content.
-ALIGN              :do full multiple alignment.
</code></pre>
</div>
<div id="perform-a-multiple-alignment-using-clustalw" class="section level3">
<h3 class="hasAnchor">
<a href="#perform-a-multiple-alignment-using-clustalw" class="anchor"></a>Perform A Multiple Alignment Using ClustalW</h3>
<p>The <code><a href="../reference/multi_aln.html">multi_aln()</a></code> function takes a fasta file storing the genes (proteins) that shall be aligned. The <code>tool</code> argument specifies the alignment tool that shall be used to perform a multiple sequence alignment (in this case <code>tool = clustalw</code>). The <code>get_aln</code> argument specifies whether or not the alignment shall be printed out to the console. In case <code>get_aln = FALSE</code>, the corresponding alignment file is stored in the <code>file.path(tempdir(),_alignment,multi_aln)</code> directory.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># in case the default execution path of clustalw runs properly on your system</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="../reference/multi_aln.html">multi_aln</a></span>(<span class="dt">file    =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/aa_seqs.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),</a>
<a class="sourceLine" id="cb2-3" data-line-number="3">          <span class="dt">tool    =</span> <span class="st">"clustalw"</span>,</a>
<a class="sourceLine" id="cb2-4" data-line-number="4">          <span class="dt">get_aln =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code> CLUSTAL 2.1 Multiple Sequence Alignments



[1] "Multiple Alignment successfully written in /var/folders/tb/
v6lh8_g505bfjytt3lvdmvw00000gn/T//RtmpOXdtWL/_alignment/multi_aln/clustalw.aln."

AAMultipleAlignment with 2 rows and 429 columns
     aln                                                                                           names               
[1] MEDQVGFGFRPNDEELVGHYLRNKIEGNTSRDVEVAISEVNICSYD...IKIPPSTNTVKQSWIVLENAQWNYLKNMIIGVLLFISVISWIILVG AT1G01010.1
[2] --------MAASEHRCVGCGFR---------------VKSLFIQYS...IFEP-------TIFLTQFGSLMQYLSYLFRTV-------------- 333554|PACid_1603...</code></pre>
<p>It is also possible to pass additional parameters to the ClustalW call:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># running clustalw using additional parameters</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="co"># details: system("clustalw2 -help")</span></a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="kw"><a href="../reference/multi_aln.html">multi_aln</a></span>(<span class="dt">file    =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/aa_seqs.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">          <span class="dt">tool    =</span> <span class="st">"clustalw"</span>,</a>
<a class="sourceLine" id="cb4-5" data-line-number="5">          <span class="dt">get_aln =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">          <span class="dt">params  =</span> <span class="st">"-PWMATRIX=BLOSUM -TYPE=PROTEIN"</span>)</a></code></pre></div>
<pre><code>CLUSTAL 2.1 Multiple Sequence Alignments


Sequence type explicitly set to Protein
Sequence format is Pearson
Sequence 1: AT1G01010.1             429 aa
Sequence 2: 333554|PACid_16033839   245 aa
Start of Pairwise alignments
Aligning...

Sequences (1:2) Aligned. Score:  2
Guide tree file created:   [/Library/Frameworks/R.framework/Versions/3.1/Resources/library/orthologr/seqs/aa_seqs.dnd]

There are 1 groups
Start of Multiple Alignment

Aligning...
Group 1:                     Delayed
Alignment Score 29

CLUSTAL-Alignment file created  [/var/folders/tb/v6lh8_g505bfjytt3lvdmvw00000gn/T//RtmpOXdtWL/_alignment/multi_aln/clustalw.aln]

[1] "Multiple Alignment successfully written in /var/folders/tb/v6lh8_g505bfjytt3lvdmvw00000gn/T//RtmpOXdtWL/_alignment/multi_aln/clustalw.aln."
AAMultipleAlignment with 2 rows and 429 columns
     aln                                                                                             names               
[1] MEDQVGFGFRPNDEELVGHYLRNKIEGNTSRDVEVAISEVNICSYD...IKIPPSTNTVKQSWIVLENAQWNYLKNMIIGVLLFISVISWIILVG AT1G01010.1
[2] --------MAASEHRCVGCGFR---------------VKSLFIQYS...IFEP-------TIFLTQFGSLMQYLSYLFRTV-------------- 333554|PACid_1603...</code></pre>
</div>
<div id="t_coffee" class="section level3">
<h3 class="hasAnchor">
<a href="#t_coffee" class="anchor"></a>T_COFFEE</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"></a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"t_coffee -version"</span>)</a></code></pre></div>
<p>In case everything is installed appropriately, you should see:</p>
<pre><code>
PROGRAM: T-COFFEE Version_11.00.8cbe486 (2014-08-12 21:55:14 - Revision 8cbe486 - Build 470)
</code></pre>
</div>
<div id="muscle" class="section level3">
<h3 class="hasAnchor">
<a href="#muscle" class="anchor"></a>MUSCLE</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"muscle -help"</span>)</a></code></pre></div>
<p>In case everything is installed appropriately, you should see:</p>
<pre><code>MUSCLE v3.8.31 by Robert C. Edgar

http://www.drive5.com/muscle
This software is donated to the public domain.
Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.


Basic usage

    muscle -in &lt;inputfile&gt; -out &lt;outputfile&gt;

Common options (for a complete list please see the User Guide):

    -in &lt;inputfile&gt;    Input file in FASTA format (default stdin)
    -out &lt;outputfile&gt;  Output alignment in FASTA format (default stdout)
    -diags             Find diagonals (faster for similar sequences)
    -maxiters &lt;n&gt;      Maximum number of iterations (integer, default 16)
    -maxhours &lt;h&gt;      Maximum time to iterate in hours (default no limit)
    -html              Write output in HTML format (default FASTA)
    -msf               Write output in GCG MSF format (default FASTA)
    -clw               Write output in CLUSTALW format (default FASTA)
    -clwstrict         As -clw, with 'CLUSTAL W (1.81)' header
    -log[a] &lt;logfile&gt;  Log to file (append if -loga, overwrite if -log)
    -quiet             Do not write progress messages to stderr
    -version           Display version information and exit

Without refinement (very fast, avg accuracy similar to T-Coffee): -maxiters 2
Fastest possible (amino acids): -maxiters 1 -diags -sv -distance1 kbit20_3
Fastest possible (nucleotides): -maxiters 1 -diags
</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co"># in case the default execution path of muscle runs properly on your system</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="kw"><a href="../reference/multi_aln.html">multi_aln</a></span>(<span class="dt">file    =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/aa_seqs.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),</a>
<a class="sourceLine" id="cb10-3" data-line-number="3">          <span class="dt">tool    =</span> <span class="st">"muscle"</span>,</a>
<a class="sourceLine" id="cb10-4" data-line-number="4">          <span class="dt">get_aln =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="clustalo" class="section level3">
<h3 class="hasAnchor">
<a href="#clustalo" class="anchor"></a>ClustalO</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1"></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"clustalo --help"</span>)</a></code></pre></div>
</div>
<div id="mafft" class="section level3">
<h3 class="hasAnchor">
<a href="#mafft" class="anchor"></a>MAFFT</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"mafft -help"</span>)</a></code></pre></div>
<p>In case everything is installed appropriately, you should see:</p>
<pre><code>
------------------------------------------------------------------------------
  MAFFT v7.187 (2014/10/02)
  http://mafft.cbrc.jp/alignment/software/
  MBE 30:772-780 (2013), NAR 30:3059-3066 (2002)
------------------------------------------------------------------------------
High speed:
  % mafft in &gt; out
  % mafft --retree 1 in &gt; out (fast)

High accuracy (for &lt;~200 sequences x &lt;~2,000 aa/nt):
  % mafft --maxiterate 1000 --localpair  in &gt; out (% linsi in &gt; out is also ok)
  % mafft --maxiterate 1000 --genafpair  in &gt; out (% einsi in &gt; out)
  % mafft --maxiterate 1000 --globalpair in &gt; out (% ginsi in &gt; out)

If unsure which option to use:
  % mafft --auto in &gt; out

--op # :         Gap opening penalty, default: 1.53
--ep # :         Offset (works like gap extension penalty), default: 0.0
--maxiterate # : Maximum number of iterative refinement, default: 0
--clustalout :   Output: clustal format, default: fasta
--reorder :      Outorder: aligned, default: input order
--quiet :        Do not report progress
--thread # :     Number of threads (if unsure, --thread -1)
</code></pre>
</div>
</div>
<div id="the-multi_aln-function-1" class="section level2">
<h2 class="hasAnchor">
<a href="#the-multi_aln-function-1" class="anchor"></a>The <code><a href="../reference/multi_aln.html">multi_aln()</a></code> function</h2>
<p>The <code><a href="../reference/multi_aln.html">multi_aln()</a></code> function is an interface function between R and common multiple sequence alignment tools. When working with this function a new folder named <code>_alignment</code> is being created and stores the multiple alignment returned by the corresponding alignment tool. The argument <code>get_aln = TRUE</code> allows to work with the multiple alignment generated by the corresponding alignment tool within the current R session.</p>
<p>This small pairwise alignment example shall illustrate how the <code><a href="../reference/multi_aln.html">multi_aln()</a></code> output can be used:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw"><a href="../reference/multi_aln.html">multi_aln</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/aa_seqs.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),</a>
<a class="sourceLine" id="cb14-2" data-line-number="2">          <span class="dt">tool =</span> <span class="st">"clustalw"</span>, <span class="dt">get_aln =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>$nb
[1] 2

$nam
[1] "AT1G01010.1"           "333554|PACid_16033839"

$seq
$seq[[1]]
[1] "medqvgfgfrpndeelvghylrnkiegntsrdvevaisevnicsydpwnlrfqskyksrdamwyffsrrennk
gnrqsrttvsgkwkltgesvevkdqwgfcsegfrgkighkrvlvfldgrypdktksdwvihefhydllpehqrtyvic
rleykgddadilsayaidptpafvpnmtssagsvvnqsrqrnsgsyntyseydsanhgqqfnensnimqqqplqgsfn
plleydfanhggqwlsdyidlqqqvpylapyenesemiwkhvieenfeflvdertsmqqhysdhrpkkpvsgvlpdds
sdtetgsmifedtssstdsvgssdepghtriddipslniieplhnykaqeqpkqqskekvissqksecewkmaedsik
ippstntvkqswivlenaqwnylknmiigvllfisviswiilvg"

$seq[[2]]
[1] "--------maasehrcvgcgfr---------------vkslfiqyspgnirlmk-------------------
---------------cgnckevadey----------iecermiifid---------------------lilhrpkvyr
hvlynainpetvniqhllwklvfvyllldsyrslllrrtdeess----------------fshssvlisikvligvls
anaafifs----------------------------------------faiaakgllnevs---rgreimlgicissy
fkifllamlvwefp----------------msvifivdilvltsnsmalkvmtestmtrciavcliahlvrfsvgqif
ep-------tifltqfgslmqylsylfrtv--------------"


$com
[1] NA

attr(,"class")
[1] "alignment"
</code></pre>
<p>Furthermore, multiple alignments are returned as follows:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="kw"><a href="../reference/multi_aln.html">multi_aln</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">'seqs/multi_aln_example.fasta'</span>, <span class="dt">package =</span> <span class="st">'orthologr'</span>),</a>
<a class="sourceLine" id="cb16-2" data-line-number="2">          <span class="dt">tool =</span> <span class="st">"clustalw"</span>, <span class="dt">get_aln =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>

 CLUSTAL 2.1 Multiple Sequence Alignments



$nb
[1] 4

$nam
[1] "AT1G01010.1|PACid_19656964"     "Thhalv10006531m|PACid_20187082"
[3] "Bra032623|PACid_22715924"       "311315|PACid_16059488"         

$seq
$seq[[1]]
[1] "-----------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
-----------------------------------------medqvgfg------------------frp
ndeelvghylrnkiegntsr----------dvevaisevnicsydp---------wnlrfqskyksrd--
-------amwyffsrre------------nnkgnrqsrttvsgkwkltgesvevkdqwgfcseg------
--frgkig-------------------------------------------hkrvlvfldgrypdktksd
wvihefhydllpehqr-----------------tyvicr--leykgddadilsayaidptpafvpnmtss
agsvvnqsrqrnsgsyntyseydsanhgqqfnensn-imqqqplqgsfnplleydfanhg-------gqw
l----------------------------sdyidlqqqvpylapye------------------------
-----------------------------------nesemiwkhvieenfeflvdertsmqqhysdhrpk
kpvsgvlp----------------ddssdtetgsmifedtssstds------------------------
---------------------------------vgssdepghtriddipslniieplhnykaqeqpkqqs
kekvissqksecewkmaeds---------ikippstntvkqswivlenaqwnylknmiigvllfisvisw
iilvg-------------------------------"

$seq[[2]]
[1] "mvmederrgdikppsywldacediscdliddlvsdfdpssvavaesvdengvnndffggidhild
siknggglpnrahingvsetnsqringnsevseaaqliagettvsvkgnvlqkcggkrdevskeegeknr
krarvcsyqrersnlsgrgqansregdrfmnrkrtrnwdeaghnkrrdgynyrrdgrdreargywerdkv
gsnelvyrsgtweadherdlkkesgrnresdekaeenkskpeehkekvveeqarryqldvleqakaknti
afletgagktliailliksihkdltsqnrkmlsvflvpkvplvyqqaevirnqtcfqvghycgemgqdfw
darrwqrefeskqvlvmtaqillnilrhsiirmeainllildechhavkkhpyslvmsefyhttpkdkrp
aifgmtaspvnlkgvssqvdcaikirnletkldstvctikdrkelekhvpmpseivveydkaatmwslhe
kikqmiaaveeaaqassrkskwqfmgardagakdelrqvygvsertesdgaanlihklrainytlaelgq
wcaykvaqsfltalqsdervnfqvdvkfqesylsevvsllqcellegaaaekavaelskpengnandeie
egelpddhvvsggehvdkvigaavadgkvtpkvqsliklllkyqhtadfraivfvervvaalvlpkvfae
lpslgfircasmighnnsqemkssqmqdtiskfrdgqvtllvatsvaeegldirqcnvvmrfdlaktvla
yiqsrgrarkpgsdyilmverenvshaaflrnarnseetlrkeaiertdlshlkdssrlisidavpgtvy
kveatgamvslnsavglihfycsqlpgdryailrpefsmvkhekpgghteyscrlqlpcnapfeilegpv
cssmrlaqqavclagckklhemgaftdmllpdkgsgqdaekadqddegepipgtarhrefypegvadvlk
gewilsgkeicessklfhlymysvrcvdsgvskdpfltevsefavlfgneldaevlsmsmdlyvaramit
kaslafrgslditesqlssikkfhvrlmsivldvdvepsttpwdpakaylfvpvadnssaepikginwel
vekitkttvwdnplqrarpdvylgtnertlggdrreygfgklrhnigfgqkshptygirgavasfdvvra
sgllpvrdalekevegdlsqgklmmadgcmvaenllgkivtaahsgkrfyvdsicydmsaetsfprkegy
lgpleyntyadyykqkygvdlsckqqplikgrgvsycknllsprfeqsgesetildktyyvflppelcvv
hplsgslvrgaqrlpsimrrvesmllavqlknlisypiptskilealtaascqetfcyeraellgdaylk
wivsrflflkypqkhegqltrmrqqmvsnlvlyqyalvkglqsyiqadrfapsrwsapgvppvydedtkd
ggssffdeeekpegnkdvfedgemedgelegdlssyrvlssktladvvealigvyyveggktaanhlmkw
igihveddpeetegsvkpvynvpesvlksidfvgleralkyeftekgllveaithasrpssgvscyqrle
fvgdavldhlitrhlfftytslppgrltdlraaavnnenfarvavkhklhlylrhgssalekqirdfvke
vltesskpgfnsfglgdckapkvlgdivesiagaifldsgkdttaawkvfqpllqpmvtpetlpmhpvre
lqercqqqaegleykasrsgntatvevfidgvqvgaaqnpqkkmaqklaarnalaalkekeaeeskkkqa
ngnaagenqddnengnkkngnqtftrqtlndiclrknwpmpsyrcvkeggpahakrftfgvrvntsdrgw
tdecigepmpsvkkakdsaailllellnktys----"

$seq[[3]]
[1] "-----------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
-----------------------------------------------mqqppqmfpmapsmpptnitteq
iqkyleenkklimaimenqnlgklaecaqyqallqknlmylaaiadaqpppstagatppp----------
---------------------------------amasqmgaphpg-------------------------
---------------------------------------------mqppsyfmqhp------qasgmaqq
appagifp----------------------prgplqfgsphqlqdp------------------------
-----------------------------------qqqhmhqqamqghmgmrpmginnnngmqhqmqqqp
etslggsaanvgirggkqdg-------------------------adgqgkddgk---------------
------------------------------------"

$seq[[4]]
[1] "-----------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
----------------------------------------------------------------------
--------------------------------------------------------mlslnmrteienlw
vfalaskfnifmqehfaslllaiaitwctltivfwstpggpawg--------------------------
--------------------------kyfftrrfsslghnrksknlipgprgfplvgsmslrsshvahqr
iasvaemsnakrlmafslgdtkvvvtchpdvakeilnssvfadrpvdetayg---lmfnramgfapngty
wrtlrrlssnhlfnpkqikrseeqrrviatqmvnafarnaksafavrdllktaslcnmmglvfg------
-----------------------------------------------------reyelesnnnveseclk
glveegydllg-----------------------tlnwtdhlpwlagldfqqirfrcsqlvpkvnlllsr
iihehyatgnfldvllslqrseklsdsdivavlwemifrgtdtvavliewvlarialhpk----------
----------------------------------vqstvhdeldr-------------------------
---------------------------------------------vvgrsrtvdesdlpsltyltamike
vlr--lhp-----------------------pgpllswarlsitdt------------------------
------------------------------------tvdgyhvpagttamvnmwaiardphvwedplefk
perfvakdgeaefsvfgsdlr---------------lapfgsgkrvcpgknlglttvsfwvatllhefew
lpsveanppdlsevlrlscemacplivnvsprrksv"


$com
[1] NA

attr(,"class")
[1] "alignment"
</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#getting-started">Getting Started</a></li>
      <li><a href="#the-multi_aln-function-1">The <code><a href="--/reference/multi_aln-html">multi_aln()</a></code> function</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Hajk-Georg Drost.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
